// @prisma/client/browser.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file should be your main import to use Prisma-related types and utilities in a browser. 
 * Use it to get access to models, enums, and input types.
 * 
 * This file does not contain a `PrismaClient` class, nor several other helpers that are intended as server-side only.
 * See `client.ts` for the standard, server-side entry point.
 *
 * 🟢 You can import this file directly.
 */

import * as Prisma from './internal/prismaNamespaceBrowser.ts'
export { Prisma }
export * as $Enums from './enums.ts'
export * from './enums.ts';
/**
 * Model ap_build_items
 * 
 */
export type ap_build_items = Prisma.ap_build_itemsModel
/**
 * Model ap_builds
 * 
 */
export type ap_builds = Prisma.ap_buildsModel
/**
 * Model champions
 * 
 */
export type champions = Prisma.championsModel
/**
 * Model doctrine_migration_versions
 * 
 */
export type doctrine_migration_versions = Prisma.doctrine_migration_versionsModel
/**
 * Model favorite_champions
 * 
 */
export type favorite_champions = Prisma.favorite_championsModel
/**
 * Model items
 * 
 */
export type items = Prisma.itemsModel
/**
 * Model player_statistics
 * 
 */
export type player_statistics = Prisma.player_statisticsModel
/**
 * Model players
 * 
 */
export type players = Prisma.playersModel
