// @prisma/client/client.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file should be your main import to use Prisma. Through it you get access to all the models, enums, and input types.
 * If you're looking for something you can import in the client-side of your application, please refer to the `browser.ts` file instead.
 * 
 * 🟢 You can import this file directly.
 */

import * as process from 'node:process'
import * as path from 'node:path'
import { fileURLToPath } from 'node:url'
globalThis['__dirname'] = path.dirname(fileURLToPath(import.meta.url))

import * as runtime from "@prisma/client/runtime/library"
import * as $Enums from "./enums.ts"
import * as $Class from "./internal/class.ts"
import * as Prisma from "./internal/prismaNamespace.ts"

export * as $Enums from './enums.ts'
export * from "./enums.ts"
/**
 * ## Prisma Client
 * 
 * Type-safe database client for TypeScript
 * @example
 * ```
 * const prisma = new PrismaClient()
 * // Fetch zero or more Ap_build_items
 * const ap_build_items = await prisma.ap_build_items.findMany()
 * ```
 * 
 * Read more in our [docs](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client).
 */
export const PrismaClient = $Class.getPrismaClientClass(__dirname)
export type PrismaClient<LogOpts extends Prisma.LogLevel = never, OmitOpts extends Prisma.PrismaClientOptions["omit"] = Prisma.PrismaClientOptions["omit"], ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = $Class.PrismaClient<LogOpts, OmitOpts, ExtArgs>
export { Prisma }


// file annotations for bundling tools to include these files
path.join(__dirname, "query_engine-windows.dll.node")
path.join(process.cwd(), "../node_modules/@prisma/client/query_engine-windows.dll.node")

/**
 * Model ap_build_items
 * 
 */
export type ap_build_items = Prisma.ap_build_itemsModel
/**
 * Model ap_builds
 * 
 */
export type ap_builds = Prisma.ap_buildsModel
/**
 * Model champions
 * 
 */
export type champions = Prisma.championsModel
/**
 * Model doctrine_migration_versions
 * 
 */
export type doctrine_migration_versions = Prisma.doctrine_migration_versionsModel
/**
 * Model favorite_champions
 * 
 */
export type favorite_champions = Prisma.favorite_championsModel
/**
 * Model items
 * 
 */
export type items = Prisma.itemsModel
/**
 * Model player_statistics
 * 
 */
export type player_statistics = Prisma.player_statisticsModel
/**
 * Model players
 * 
 */
export type players = Prisma.playersModel
