// @prisma/client/models/ap_build_items.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `ap_build_items` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/library"
import type * as $Enums from "../enums.ts"
import type * as Prisma from "../internal/prismaNamespace.ts"

/**
 * Model ap_build_items
 * 
 */
export type ap_build_itemsModel = runtime.Types.Result.DefaultSelection<Prisma.$ap_build_itemsPayload>

export type AggregateAp_build_items = {
  _count: Ap_build_itemsCountAggregateOutputType | null
  _avg: Ap_build_itemsAvgAggregateOutputType | null
  _sum: Ap_build_itemsSumAggregateOutputType | null
  _min: Ap_build_itemsMinAggregateOutputType | null
  _max: Ap_build_itemsMaxAggregateOutputType | null
}

export type Ap_build_itemsAvgAggregateOutputType = {
  ap_build_id: number | null
  item_id: number | null
}

export type Ap_build_itemsSumAggregateOutputType = {
  ap_build_id: number | null
  item_id: number | null
}

export type Ap_build_itemsMinAggregateOutputType = {
  ap_build_id: number | null
  item_id: number | null
}

export type Ap_build_itemsMaxAggregateOutputType = {
  ap_build_id: number | null
  item_id: number | null
}

export type Ap_build_itemsCountAggregateOutputType = {
  ap_build_id: number
  item_id: number
  _all: number
}


export type Ap_build_itemsAvgAggregateInputType = {
  ap_build_id?: true
  item_id?: true
}

export type Ap_build_itemsSumAggregateInputType = {
  ap_build_id?: true
  item_id?: true
}

export type Ap_build_itemsMinAggregateInputType = {
  ap_build_id?: true
  item_id?: true
}

export type Ap_build_itemsMaxAggregateInputType = {
  ap_build_id?: true
  item_id?: true
}

export type Ap_build_itemsCountAggregateInputType = {
  ap_build_id?: true
  item_id?: true
  _all?: true
}

export type Ap_build_itemsAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which ap_build_items to aggregate.
   */
  where?: Prisma.ap_build_itemsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of ap_build_items to fetch.
   */
  orderBy?: Prisma.ap_build_itemsOrderByWithRelationInput | Prisma.ap_build_itemsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.ap_build_itemsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` ap_build_items from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` ap_build_items.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned ap_build_items
  **/
  _count?: true | Ap_build_itemsCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to average
  **/
  _avg?: Ap_build_itemsAvgAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to sum
  **/
  _sum?: Ap_build_itemsSumAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: Ap_build_itemsMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: Ap_build_itemsMaxAggregateInputType
}

export type GetAp_build_itemsAggregateType<T extends Ap_build_itemsAggregateArgs> = {
      [P in keyof T & keyof AggregateAp_build_items]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateAp_build_items[P]>
    : Prisma.GetScalarType<T[P], AggregateAp_build_items[P]>
}




export type ap_build_itemsGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.ap_build_itemsWhereInput
  orderBy?: Prisma.ap_build_itemsOrderByWithAggregationInput | Prisma.ap_build_itemsOrderByWithAggregationInput[]
  by: Prisma.Ap_build_itemsScalarFieldEnum[] | Prisma.Ap_build_itemsScalarFieldEnum
  having?: Prisma.ap_build_itemsScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: Ap_build_itemsCountAggregateInputType | true
  _avg?: Ap_build_itemsAvgAggregateInputType
  _sum?: Ap_build_itemsSumAggregateInputType
  _min?: Ap_build_itemsMinAggregateInputType
  _max?: Ap_build_itemsMaxAggregateInputType
}

export type Ap_build_itemsGroupByOutputType = {
  ap_build_id: number
  item_id: number
  _count: Ap_build_itemsCountAggregateOutputType | null
  _avg: Ap_build_itemsAvgAggregateOutputType | null
  _sum: Ap_build_itemsSumAggregateOutputType | null
  _min: Ap_build_itemsMinAggregateOutputType | null
  _max: Ap_build_itemsMaxAggregateOutputType | null
}

type GetAp_build_itemsGroupByPayload<T extends ap_build_itemsGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<Ap_build_itemsGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof Ap_build_itemsGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], Ap_build_itemsGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], Ap_build_itemsGroupByOutputType[P]>
      }
    >
  >



export type ap_build_itemsWhereInput = {
  AND?: Prisma.ap_build_itemsWhereInput | Prisma.ap_build_itemsWhereInput[]
  OR?: Prisma.ap_build_itemsWhereInput[]
  NOT?: Prisma.ap_build_itemsWhereInput | Prisma.ap_build_itemsWhereInput[]
  ap_build_id?: Prisma.IntFilter<"ap_build_items"> | number
  item_id?: Prisma.IntFilter<"ap_build_items"> | number
  ap_builds?: Prisma.XOR<Prisma.Ap_buildsScalarRelationFilter, Prisma.ap_buildsWhereInput>
  items?: Prisma.XOR<Prisma.ItemsScalarRelationFilter, Prisma.itemsWhereInput>
}

export type ap_build_itemsOrderByWithRelationInput = {
  ap_build_id?: Prisma.SortOrder
  item_id?: Prisma.SortOrder
  ap_builds?: Prisma.ap_buildsOrderByWithRelationInput
  items?: Prisma.itemsOrderByWithRelationInput
}

export type ap_build_itemsWhereUniqueInput = Prisma.AtLeast<{
  ap_build_id_item_id?: Prisma.ap_build_itemsAp_build_idItem_idCompoundUniqueInput
  AND?: Prisma.ap_build_itemsWhereInput | Prisma.ap_build_itemsWhereInput[]
  OR?: Prisma.ap_build_itemsWhereInput[]
  NOT?: Prisma.ap_build_itemsWhereInput | Prisma.ap_build_itemsWhereInput[]
  ap_build_id?: Prisma.IntFilter<"ap_build_items"> | number
  item_id?: Prisma.IntFilter<"ap_build_items"> | number
  ap_builds?: Prisma.XOR<Prisma.Ap_buildsScalarRelationFilter, Prisma.ap_buildsWhereInput>
  items?: Prisma.XOR<Prisma.ItemsScalarRelationFilter, Prisma.itemsWhereInput>
}, "ap_build_id_item_id">

export type ap_build_itemsOrderByWithAggregationInput = {
  ap_build_id?: Prisma.SortOrder
  item_id?: Prisma.SortOrder
  _count?: Prisma.ap_build_itemsCountOrderByAggregateInput
  _avg?: Prisma.ap_build_itemsAvgOrderByAggregateInput
  _max?: Prisma.ap_build_itemsMaxOrderByAggregateInput
  _min?: Prisma.ap_build_itemsMinOrderByAggregateInput
  _sum?: Prisma.ap_build_itemsSumOrderByAggregateInput
}

export type ap_build_itemsScalarWhereWithAggregatesInput = {
  AND?: Prisma.ap_build_itemsScalarWhereWithAggregatesInput | Prisma.ap_build_itemsScalarWhereWithAggregatesInput[]
  OR?: Prisma.ap_build_itemsScalarWhereWithAggregatesInput[]
  NOT?: Prisma.ap_build_itemsScalarWhereWithAggregatesInput | Prisma.ap_build_itemsScalarWhereWithAggregatesInput[]
  ap_build_id?: Prisma.IntWithAggregatesFilter<"ap_build_items"> | number
  item_id?: Prisma.IntWithAggregatesFilter<"ap_build_items"> | number
}

export type ap_build_itemsCreateInput = {
  ap_builds: Prisma.ap_buildsCreateNestedOneWithoutAp_build_itemsInput
  items: Prisma.itemsCreateNestedOneWithoutAp_build_itemsInput
}

export type ap_build_itemsUncheckedCreateInput = {
  ap_build_id: number
  item_id: number
}

export type ap_build_itemsUpdateInput = {
  ap_builds?: Prisma.ap_buildsUpdateOneRequiredWithoutAp_build_itemsNestedInput
  items?: Prisma.itemsUpdateOneRequiredWithoutAp_build_itemsNestedInput
}

export type ap_build_itemsUncheckedUpdateInput = {
  ap_build_id?: Prisma.IntFieldUpdateOperationsInput | number
  item_id?: Prisma.IntFieldUpdateOperationsInput | number
}

export type ap_build_itemsCreateManyInput = {
  ap_build_id: number
  item_id: number
}

export type ap_build_itemsUpdateManyMutationInput = {

}

export type ap_build_itemsUncheckedUpdateManyInput = {
  ap_build_id?: Prisma.IntFieldUpdateOperationsInput | number
  item_id?: Prisma.IntFieldUpdateOperationsInput | number
}

export type ap_build_itemsAp_build_idItem_idCompoundUniqueInput = {
  ap_build_id: number
  item_id: number
}

export type ap_build_itemsCountOrderByAggregateInput = {
  ap_build_id?: Prisma.SortOrder
  item_id?: Prisma.SortOrder
}

export type ap_build_itemsAvgOrderByAggregateInput = {
  ap_build_id?: Prisma.SortOrder
  item_id?: Prisma.SortOrder
}

export type ap_build_itemsMaxOrderByAggregateInput = {
  ap_build_id?: Prisma.SortOrder
  item_id?: Prisma.SortOrder
}

export type ap_build_itemsMinOrderByAggregateInput = {
  ap_build_id?: Prisma.SortOrder
  item_id?: Prisma.SortOrder
}

export type ap_build_itemsSumOrderByAggregateInput = {
  ap_build_id?: Prisma.SortOrder
  item_id?: Prisma.SortOrder
}

export type Ap_build_itemsListRelationFilter = {
  every?: Prisma.ap_build_itemsWhereInput
  some?: Prisma.ap_build_itemsWhereInput
  none?: Prisma.ap_build_itemsWhereInput
}

export type ap_build_itemsOrderByRelationAggregateInput = {
  _count?: Prisma.SortOrder
}

export type IntFieldUpdateOperationsInput = {
  set?: number
  increment?: number
  decrement?: number
  multiply?: number
  divide?: number
}

export type ap_build_itemsCreateNestedManyWithoutAp_buildsInput = {
  create?: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutAp_buildsInput, Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput> | Prisma.ap_build_itemsCreateWithoutAp_buildsInput[] | Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput[]
  connectOrCreate?: Prisma.ap_build_itemsCreateOrConnectWithoutAp_buildsInput | Prisma.ap_build_itemsCreateOrConnectWithoutAp_buildsInput[]
  createMany?: Prisma.ap_build_itemsCreateManyAp_buildsInputEnvelope
  connect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
}

export type ap_build_itemsUncheckedCreateNestedManyWithoutAp_buildsInput = {
  create?: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutAp_buildsInput, Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput> | Prisma.ap_build_itemsCreateWithoutAp_buildsInput[] | Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput[]
  connectOrCreate?: Prisma.ap_build_itemsCreateOrConnectWithoutAp_buildsInput | Prisma.ap_build_itemsCreateOrConnectWithoutAp_buildsInput[]
  createMany?: Prisma.ap_build_itemsCreateManyAp_buildsInputEnvelope
  connect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
}

export type ap_build_itemsUpdateManyWithoutAp_buildsNestedInput = {
  create?: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutAp_buildsInput, Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput> | Prisma.ap_build_itemsCreateWithoutAp_buildsInput[] | Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput[]
  connectOrCreate?: Prisma.ap_build_itemsCreateOrConnectWithoutAp_buildsInput | Prisma.ap_build_itemsCreateOrConnectWithoutAp_buildsInput[]
  upsert?: Prisma.ap_build_itemsUpsertWithWhereUniqueWithoutAp_buildsInput | Prisma.ap_build_itemsUpsertWithWhereUniqueWithoutAp_buildsInput[]
  createMany?: Prisma.ap_build_itemsCreateManyAp_buildsInputEnvelope
  set?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  disconnect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  delete?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  connect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  update?: Prisma.ap_build_itemsUpdateWithWhereUniqueWithoutAp_buildsInput | Prisma.ap_build_itemsUpdateWithWhereUniqueWithoutAp_buildsInput[]
  updateMany?: Prisma.ap_build_itemsUpdateManyWithWhereWithoutAp_buildsInput | Prisma.ap_build_itemsUpdateManyWithWhereWithoutAp_buildsInput[]
  deleteMany?: Prisma.ap_build_itemsScalarWhereInput | Prisma.ap_build_itemsScalarWhereInput[]
}

export type ap_build_itemsUncheckedUpdateManyWithoutAp_buildsNestedInput = {
  create?: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutAp_buildsInput, Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput> | Prisma.ap_build_itemsCreateWithoutAp_buildsInput[] | Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput[]
  connectOrCreate?: Prisma.ap_build_itemsCreateOrConnectWithoutAp_buildsInput | Prisma.ap_build_itemsCreateOrConnectWithoutAp_buildsInput[]
  upsert?: Prisma.ap_build_itemsUpsertWithWhereUniqueWithoutAp_buildsInput | Prisma.ap_build_itemsUpsertWithWhereUniqueWithoutAp_buildsInput[]
  createMany?: Prisma.ap_build_itemsCreateManyAp_buildsInputEnvelope
  set?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  disconnect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  delete?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  connect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  update?: Prisma.ap_build_itemsUpdateWithWhereUniqueWithoutAp_buildsInput | Prisma.ap_build_itemsUpdateWithWhereUniqueWithoutAp_buildsInput[]
  updateMany?: Prisma.ap_build_itemsUpdateManyWithWhereWithoutAp_buildsInput | Prisma.ap_build_itemsUpdateManyWithWhereWithoutAp_buildsInput[]
  deleteMany?: Prisma.ap_build_itemsScalarWhereInput | Prisma.ap_build_itemsScalarWhereInput[]
}

export type ap_build_itemsCreateNestedManyWithoutItemsInput = {
  create?: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutItemsInput, Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput> | Prisma.ap_build_itemsCreateWithoutItemsInput[] | Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput[]
  connectOrCreate?: Prisma.ap_build_itemsCreateOrConnectWithoutItemsInput | Prisma.ap_build_itemsCreateOrConnectWithoutItemsInput[]
  createMany?: Prisma.ap_build_itemsCreateManyItemsInputEnvelope
  connect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
}

export type ap_build_itemsUncheckedCreateNestedManyWithoutItemsInput = {
  create?: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutItemsInput, Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput> | Prisma.ap_build_itemsCreateWithoutItemsInput[] | Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput[]
  connectOrCreate?: Prisma.ap_build_itemsCreateOrConnectWithoutItemsInput | Prisma.ap_build_itemsCreateOrConnectWithoutItemsInput[]
  createMany?: Prisma.ap_build_itemsCreateManyItemsInputEnvelope
  connect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
}

export type ap_build_itemsUpdateManyWithoutItemsNestedInput = {
  create?: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutItemsInput, Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput> | Prisma.ap_build_itemsCreateWithoutItemsInput[] | Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput[]
  connectOrCreate?: Prisma.ap_build_itemsCreateOrConnectWithoutItemsInput | Prisma.ap_build_itemsCreateOrConnectWithoutItemsInput[]
  upsert?: Prisma.ap_build_itemsUpsertWithWhereUniqueWithoutItemsInput | Prisma.ap_build_itemsUpsertWithWhereUniqueWithoutItemsInput[]
  createMany?: Prisma.ap_build_itemsCreateManyItemsInputEnvelope
  set?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  disconnect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  delete?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  connect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  update?: Prisma.ap_build_itemsUpdateWithWhereUniqueWithoutItemsInput | Prisma.ap_build_itemsUpdateWithWhereUniqueWithoutItemsInput[]
  updateMany?: Prisma.ap_build_itemsUpdateManyWithWhereWithoutItemsInput | Prisma.ap_build_itemsUpdateManyWithWhereWithoutItemsInput[]
  deleteMany?: Prisma.ap_build_itemsScalarWhereInput | Prisma.ap_build_itemsScalarWhereInput[]
}

export type ap_build_itemsUncheckedUpdateManyWithoutItemsNestedInput = {
  create?: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutItemsInput, Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput> | Prisma.ap_build_itemsCreateWithoutItemsInput[] | Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput[]
  connectOrCreate?: Prisma.ap_build_itemsCreateOrConnectWithoutItemsInput | Prisma.ap_build_itemsCreateOrConnectWithoutItemsInput[]
  upsert?: Prisma.ap_build_itemsUpsertWithWhereUniqueWithoutItemsInput | Prisma.ap_build_itemsUpsertWithWhereUniqueWithoutItemsInput[]
  createMany?: Prisma.ap_build_itemsCreateManyItemsInputEnvelope
  set?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  disconnect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  delete?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  connect?: Prisma.ap_build_itemsWhereUniqueInput | Prisma.ap_build_itemsWhereUniqueInput[]
  update?: Prisma.ap_build_itemsUpdateWithWhereUniqueWithoutItemsInput | Prisma.ap_build_itemsUpdateWithWhereUniqueWithoutItemsInput[]
  updateMany?: Prisma.ap_build_itemsUpdateManyWithWhereWithoutItemsInput | Prisma.ap_build_itemsUpdateManyWithWhereWithoutItemsInput[]
  deleteMany?: Prisma.ap_build_itemsScalarWhereInput | Prisma.ap_build_itemsScalarWhereInput[]
}

export type ap_build_itemsCreateWithoutAp_buildsInput = {
  items: Prisma.itemsCreateNestedOneWithoutAp_build_itemsInput
}

export type ap_build_itemsUncheckedCreateWithoutAp_buildsInput = {
  item_id: number
}

export type ap_build_itemsCreateOrConnectWithoutAp_buildsInput = {
  where: Prisma.ap_build_itemsWhereUniqueInput
  create: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutAp_buildsInput, Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput>
}

export type ap_build_itemsCreateManyAp_buildsInputEnvelope = {
  data: Prisma.ap_build_itemsCreateManyAp_buildsInput | Prisma.ap_build_itemsCreateManyAp_buildsInput[]
  skipDuplicates?: boolean
}

export type ap_build_itemsUpsertWithWhereUniqueWithoutAp_buildsInput = {
  where: Prisma.ap_build_itemsWhereUniqueInput
  update: Prisma.XOR<Prisma.ap_build_itemsUpdateWithoutAp_buildsInput, Prisma.ap_build_itemsUncheckedUpdateWithoutAp_buildsInput>
  create: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutAp_buildsInput, Prisma.ap_build_itemsUncheckedCreateWithoutAp_buildsInput>
}

export type ap_build_itemsUpdateWithWhereUniqueWithoutAp_buildsInput = {
  where: Prisma.ap_build_itemsWhereUniqueInput
  data: Prisma.XOR<Prisma.ap_build_itemsUpdateWithoutAp_buildsInput, Prisma.ap_build_itemsUncheckedUpdateWithoutAp_buildsInput>
}

export type ap_build_itemsUpdateManyWithWhereWithoutAp_buildsInput = {
  where: Prisma.ap_build_itemsScalarWhereInput
  data: Prisma.XOR<Prisma.ap_build_itemsUpdateManyMutationInput, Prisma.ap_build_itemsUncheckedUpdateManyWithoutAp_buildsInput>
}

export type ap_build_itemsScalarWhereInput = {
  AND?: Prisma.ap_build_itemsScalarWhereInput | Prisma.ap_build_itemsScalarWhereInput[]
  OR?: Prisma.ap_build_itemsScalarWhereInput[]
  NOT?: Prisma.ap_build_itemsScalarWhereInput | Prisma.ap_build_itemsScalarWhereInput[]
  ap_build_id?: Prisma.IntFilter<"ap_build_items"> | number
  item_id?: Prisma.IntFilter<"ap_build_items"> | number
}

export type ap_build_itemsCreateWithoutItemsInput = {
  ap_builds: Prisma.ap_buildsCreateNestedOneWithoutAp_build_itemsInput
}

export type ap_build_itemsUncheckedCreateWithoutItemsInput = {
  ap_build_id: number
}

export type ap_build_itemsCreateOrConnectWithoutItemsInput = {
  where: Prisma.ap_build_itemsWhereUniqueInput
  create: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutItemsInput, Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput>
}

export type ap_build_itemsCreateManyItemsInputEnvelope = {
  data: Prisma.ap_build_itemsCreateManyItemsInput | Prisma.ap_build_itemsCreateManyItemsInput[]
  skipDuplicates?: boolean
}

export type ap_build_itemsUpsertWithWhereUniqueWithoutItemsInput = {
  where: Prisma.ap_build_itemsWhereUniqueInput
  update: Prisma.XOR<Prisma.ap_build_itemsUpdateWithoutItemsInput, Prisma.ap_build_itemsUncheckedUpdateWithoutItemsInput>
  create: Prisma.XOR<Prisma.ap_build_itemsCreateWithoutItemsInput, Prisma.ap_build_itemsUncheckedCreateWithoutItemsInput>
}

export type ap_build_itemsUpdateWithWhereUniqueWithoutItemsInput = {
  where: Prisma.ap_build_itemsWhereUniqueInput
  data: Prisma.XOR<Prisma.ap_build_itemsUpdateWithoutItemsInput, Prisma.ap_build_itemsUncheckedUpdateWithoutItemsInput>
}

export type ap_build_itemsUpdateManyWithWhereWithoutItemsInput = {
  where: Prisma.ap_build_itemsScalarWhereInput
  data: Prisma.XOR<Prisma.ap_build_itemsUpdateManyMutationInput, Prisma.ap_build_itemsUncheckedUpdateManyWithoutItemsInput>
}

export type ap_build_itemsCreateManyAp_buildsInput = {
  item_id: number
}

export type ap_build_itemsUpdateWithoutAp_buildsInput = {
  items?: Prisma.itemsUpdateOneRequiredWithoutAp_build_itemsNestedInput
}

export type ap_build_itemsUncheckedUpdateWithoutAp_buildsInput = {
  item_id?: Prisma.IntFieldUpdateOperationsInput | number
}

export type ap_build_itemsUncheckedUpdateManyWithoutAp_buildsInput = {
  item_id?: Prisma.IntFieldUpdateOperationsInput | number
}

export type ap_build_itemsCreateManyItemsInput = {
  ap_build_id: number
}

export type ap_build_itemsUpdateWithoutItemsInput = {
  ap_builds?: Prisma.ap_buildsUpdateOneRequiredWithoutAp_build_itemsNestedInput
}

export type ap_build_itemsUncheckedUpdateWithoutItemsInput = {
  ap_build_id?: Prisma.IntFieldUpdateOperationsInput | number
}

export type ap_build_itemsUncheckedUpdateManyWithoutItemsInput = {
  ap_build_id?: Prisma.IntFieldUpdateOperationsInput | number
}



export type ap_build_itemsSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  ap_build_id?: boolean
  item_id?: boolean
  ap_builds?: boolean | Prisma.ap_buildsDefaultArgs<ExtArgs>
  items?: boolean | Prisma.itemsDefaultArgs<ExtArgs>
}, ExtArgs["result"]["ap_build_items"]>

export type ap_build_itemsSelectCreateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  ap_build_id?: boolean
  item_id?: boolean
  ap_builds?: boolean | Prisma.ap_buildsDefaultArgs<ExtArgs>
  items?: boolean | Prisma.itemsDefaultArgs<ExtArgs>
}, ExtArgs["result"]["ap_build_items"]>

export type ap_build_itemsSelectUpdateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  ap_build_id?: boolean
  item_id?: boolean
  ap_builds?: boolean | Prisma.ap_buildsDefaultArgs<ExtArgs>
  items?: boolean | Prisma.itemsDefaultArgs<ExtArgs>
}, ExtArgs["result"]["ap_build_items"]>

export type ap_build_itemsSelectScalar = {
  ap_build_id?: boolean
  item_id?: boolean
}

export type ap_build_itemsOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"ap_build_id" | "item_id", ExtArgs["result"]["ap_build_items"]>
export type ap_build_itemsInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  ap_builds?: boolean | Prisma.ap_buildsDefaultArgs<ExtArgs>
  items?: boolean | Prisma.itemsDefaultArgs<ExtArgs>
}
export type ap_build_itemsIncludeCreateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  ap_builds?: boolean | Prisma.ap_buildsDefaultArgs<ExtArgs>
  items?: boolean | Prisma.itemsDefaultArgs<ExtArgs>
}
export type ap_build_itemsIncludeUpdateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  ap_builds?: boolean | Prisma.ap_buildsDefaultArgs<ExtArgs>
  items?: boolean | Prisma.itemsDefaultArgs<ExtArgs>
}

export type $ap_build_itemsPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "ap_build_items"
  objects: {
    ap_builds: Prisma.$ap_buildsPayload<ExtArgs>
    items: Prisma.$itemsPayload<ExtArgs>
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    ap_build_id: number
    item_id: number
  }, ExtArgs["result"]["ap_build_items"]>
  composites: {}
}

export type ap_build_itemsGetPayload<S extends boolean | null | undefined | ap_build_itemsDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload, S>

export type ap_build_itemsCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<ap_build_itemsFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: Ap_build_itemsCountAggregateInputType | true
  }

export interface ap_build_itemsDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['ap_build_items'], meta: { name: 'ap_build_items' } }
  /**
   * Find zero or one Ap_build_items that matches the filter.
   * @param {ap_build_itemsFindUniqueArgs} args - Arguments to find a Ap_build_items
   * @example
   * // Get one Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends ap_build_itemsFindUniqueArgs>(args: Prisma.SelectSubset<T, ap_build_itemsFindUniqueArgs<ExtArgs>>): Prisma.Prisma__ap_build_itemsClient<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one Ap_build_items that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {ap_build_itemsFindUniqueOrThrowArgs} args - Arguments to find a Ap_build_items
   * @example
   * // Get one Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends ap_build_itemsFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, ap_build_itemsFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__ap_build_itemsClient<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Ap_build_items that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ap_build_itemsFindFirstArgs} args - Arguments to find a Ap_build_items
   * @example
   * // Get one Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends ap_build_itemsFindFirstArgs>(args?: Prisma.SelectSubset<T, ap_build_itemsFindFirstArgs<ExtArgs>>): Prisma.Prisma__ap_build_itemsClient<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Ap_build_items that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ap_build_itemsFindFirstOrThrowArgs} args - Arguments to find a Ap_build_items
   * @example
   * // Get one Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends ap_build_itemsFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, ap_build_itemsFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__ap_build_itemsClient<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Ap_build_items that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ap_build_itemsFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.findMany()
   * 
   * // Get first 10 Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.findMany({ take: 10 })
   * 
   * // Only select the `ap_build_id`
   * const ap_build_itemsWithAp_build_idOnly = await prisma.ap_build_items.findMany({ select: { ap_build_id: true } })
   * 
   */
  findMany<T extends ap_build_itemsFindManyArgs>(args?: Prisma.SelectSubset<T, ap_build_itemsFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a Ap_build_items.
   * @param {ap_build_itemsCreateArgs} args - Arguments to create a Ap_build_items.
   * @example
   * // Create one Ap_build_items
   * const Ap_build_items = await prisma.ap_build_items.create({
   *   data: {
   *     // ... data to create a Ap_build_items
   *   }
   * })
   * 
   */
  create<T extends ap_build_itemsCreateArgs>(args: Prisma.SelectSubset<T, ap_build_itemsCreateArgs<ExtArgs>>): Prisma.Prisma__ap_build_itemsClient<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many Ap_build_items.
   * @param {ap_build_itemsCreateManyArgs} args - Arguments to create many Ap_build_items.
   * @example
   * // Create many Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends ap_build_itemsCreateManyArgs>(args?: Prisma.SelectSubset<T, ap_build_itemsCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create many Ap_build_items and returns the data saved in the database.
   * @param {ap_build_itemsCreateManyAndReturnArgs} args - Arguments to create many Ap_build_items.
   * @example
   * // Create many Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.createManyAndReturn({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * 
   * // Create many Ap_build_items and only return the `ap_build_id`
   * const ap_build_itemsWithAp_build_idOnly = await prisma.ap_build_items.createManyAndReturn({
   *   select: { ap_build_id: true },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * 
   */
  createManyAndReturn<T extends ap_build_itemsCreateManyAndReturnArgs>(args?: Prisma.SelectSubset<T, ap_build_itemsCreateManyAndReturnArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "createManyAndReturn", GlobalOmitOptions>>

  /**
   * Delete a Ap_build_items.
   * @param {ap_build_itemsDeleteArgs} args - Arguments to delete one Ap_build_items.
   * @example
   * // Delete one Ap_build_items
   * const Ap_build_items = await prisma.ap_build_items.delete({
   *   where: {
   *     // ... filter to delete one Ap_build_items
   *   }
   * })
   * 
   */
  delete<T extends ap_build_itemsDeleteArgs>(args: Prisma.SelectSubset<T, ap_build_itemsDeleteArgs<ExtArgs>>): Prisma.Prisma__ap_build_itemsClient<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one Ap_build_items.
   * @param {ap_build_itemsUpdateArgs} args - Arguments to update one Ap_build_items.
   * @example
   * // Update one Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends ap_build_itemsUpdateArgs>(args: Prisma.SelectSubset<T, ap_build_itemsUpdateArgs<ExtArgs>>): Prisma.Prisma__ap_build_itemsClient<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more Ap_build_items.
   * @param {ap_build_itemsDeleteManyArgs} args - Arguments to filter Ap_build_items to delete.
   * @example
   * // Delete a few Ap_build_items
   * const { count } = await prisma.ap_build_items.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends ap_build_itemsDeleteManyArgs>(args?: Prisma.SelectSubset<T, ap_build_itemsDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Ap_build_items.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ap_build_itemsUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends ap_build_itemsUpdateManyArgs>(args: Prisma.SelectSubset<T, ap_build_itemsUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Ap_build_items and returns the data updated in the database.
   * @param {ap_build_itemsUpdateManyAndReturnArgs} args - Arguments to update many Ap_build_items.
   * @example
   * // Update many Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.updateManyAndReturn({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * 
   * // Update zero or more Ap_build_items and only return the `ap_build_id`
   * const ap_build_itemsWithAp_build_idOnly = await prisma.ap_build_items.updateManyAndReturn({
   *   select: { ap_build_id: true },
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * 
   */
  updateManyAndReturn<T extends ap_build_itemsUpdateManyAndReturnArgs>(args: Prisma.SelectSubset<T, ap_build_itemsUpdateManyAndReturnArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "updateManyAndReturn", GlobalOmitOptions>>

  /**
   * Create or update one Ap_build_items.
   * @param {ap_build_itemsUpsertArgs} args - Arguments to update or create a Ap_build_items.
   * @example
   * // Update or create a Ap_build_items
   * const ap_build_items = await prisma.ap_build_items.upsert({
   *   create: {
   *     // ... data to create a Ap_build_items
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the Ap_build_items we want to update
   *   }
   * })
   */
  upsert<T extends ap_build_itemsUpsertArgs>(args: Prisma.SelectSubset<T, ap_build_itemsUpsertArgs<ExtArgs>>): Prisma.Prisma__ap_build_itemsClient<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>


  /**
   * Count the number of Ap_build_items.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ap_build_itemsCountArgs} args - Arguments to filter Ap_build_items to count.
   * @example
   * // Count the number of Ap_build_items
   * const count = await prisma.ap_build_items.count({
   *   where: {
   *     // ... the filter for the Ap_build_items we want to count
   *   }
   * })
  **/
  count<T extends ap_build_itemsCountArgs>(
    args?: Prisma.Subset<T, ap_build_itemsCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], Ap_build_itemsCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a Ap_build_items.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {Ap_build_itemsAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends Ap_build_itemsAggregateArgs>(args: Prisma.Subset<T, Ap_build_itemsAggregateArgs>): Prisma.PrismaPromise<GetAp_build_itemsAggregateType<T>>

  /**
   * Group by Ap_build_items.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ap_build_itemsGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends ap_build_itemsGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: ap_build_itemsGroupByArgs['orderBy'] }
      : { orderBy?: ap_build_itemsGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, ap_build_itemsGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetAp_build_itemsGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the ap_build_items model
 */
readonly fields: ap_build_itemsFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for ap_build_items.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__ap_build_itemsClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  ap_builds<T extends Prisma.ap_buildsDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.ap_buildsDefaultArgs<ExtArgs>>): Prisma.Prisma__ap_buildsClient<runtime.Types.Result.GetResult<Prisma.$ap_buildsPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  items<T extends Prisma.itemsDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.itemsDefaultArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the ap_build_items model
 */
export interface ap_build_itemsFieldRefs {
  readonly ap_build_id: Prisma.FieldRef<"ap_build_items", 'Int'>
  readonly item_id: Prisma.FieldRef<"ap_build_items", 'Int'>
}
    

// Custom InputTypes
/**
 * ap_build_items findUnique
 */
export type ap_build_itemsFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * Filter, which ap_build_items to fetch.
   */
  where: Prisma.ap_build_itemsWhereUniqueInput
}

/**
 * ap_build_items findUniqueOrThrow
 */
export type ap_build_itemsFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * Filter, which ap_build_items to fetch.
   */
  where: Prisma.ap_build_itemsWhereUniqueInput
}

/**
 * ap_build_items findFirst
 */
export type ap_build_itemsFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * Filter, which ap_build_items to fetch.
   */
  where?: Prisma.ap_build_itemsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of ap_build_items to fetch.
   */
  orderBy?: Prisma.ap_build_itemsOrderByWithRelationInput | Prisma.ap_build_itemsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for ap_build_items.
   */
  cursor?: Prisma.ap_build_itemsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` ap_build_items from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` ap_build_items.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of ap_build_items.
   */
  distinct?: Prisma.Ap_build_itemsScalarFieldEnum | Prisma.Ap_build_itemsScalarFieldEnum[]
}

/**
 * ap_build_items findFirstOrThrow
 */
export type ap_build_itemsFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * Filter, which ap_build_items to fetch.
   */
  where?: Prisma.ap_build_itemsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of ap_build_items to fetch.
   */
  orderBy?: Prisma.ap_build_itemsOrderByWithRelationInput | Prisma.ap_build_itemsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for ap_build_items.
   */
  cursor?: Prisma.ap_build_itemsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` ap_build_items from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` ap_build_items.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of ap_build_items.
   */
  distinct?: Prisma.Ap_build_itemsScalarFieldEnum | Prisma.Ap_build_itemsScalarFieldEnum[]
}

/**
 * ap_build_items findMany
 */
export type ap_build_itemsFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * Filter, which ap_build_items to fetch.
   */
  where?: Prisma.ap_build_itemsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of ap_build_items to fetch.
   */
  orderBy?: Prisma.ap_build_itemsOrderByWithRelationInput | Prisma.ap_build_itemsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing ap_build_items.
   */
  cursor?: Prisma.ap_build_itemsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` ap_build_items from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` ap_build_items.
   */
  skip?: number
  distinct?: Prisma.Ap_build_itemsScalarFieldEnum | Prisma.Ap_build_itemsScalarFieldEnum[]
}

/**
 * ap_build_items create
 */
export type ap_build_itemsCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * The data needed to create a ap_build_items.
   */
  data: Prisma.XOR<Prisma.ap_build_itemsCreateInput, Prisma.ap_build_itemsUncheckedCreateInput>
}

/**
 * ap_build_items createMany
 */
export type ap_build_itemsCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many ap_build_items.
   */
  data: Prisma.ap_build_itemsCreateManyInput | Prisma.ap_build_itemsCreateManyInput[]
  skipDuplicates?: boolean
}

/**
 * ap_build_items createManyAndReturn
 */
export type ap_build_itemsCreateManyAndReturnArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelectCreateManyAndReturn<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * The data used to create many ap_build_items.
   */
  data: Prisma.ap_build_itemsCreateManyInput | Prisma.ap_build_itemsCreateManyInput[]
  skipDuplicates?: boolean
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsIncludeCreateManyAndReturn<ExtArgs> | null
}

/**
 * ap_build_items update
 */
export type ap_build_itemsUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * The data needed to update a ap_build_items.
   */
  data: Prisma.XOR<Prisma.ap_build_itemsUpdateInput, Prisma.ap_build_itemsUncheckedUpdateInput>
  /**
   * Choose, which ap_build_items to update.
   */
  where: Prisma.ap_build_itemsWhereUniqueInput
}

/**
 * ap_build_items updateMany
 */
export type ap_build_itemsUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update ap_build_items.
   */
  data: Prisma.XOR<Prisma.ap_build_itemsUpdateManyMutationInput, Prisma.ap_build_itemsUncheckedUpdateManyInput>
  /**
   * Filter which ap_build_items to update
   */
  where?: Prisma.ap_build_itemsWhereInput
  /**
   * Limit how many ap_build_items to update.
   */
  limit?: number
}

/**
 * ap_build_items updateManyAndReturn
 */
export type ap_build_itemsUpdateManyAndReturnArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelectUpdateManyAndReturn<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * The data used to update ap_build_items.
   */
  data: Prisma.XOR<Prisma.ap_build_itemsUpdateManyMutationInput, Prisma.ap_build_itemsUncheckedUpdateManyInput>
  /**
   * Filter which ap_build_items to update
   */
  where?: Prisma.ap_build_itemsWhereInput
  /**
   * Limit how many ap_build_items to update.
   */
  limit?: number
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsIncludeUpdateManyAndReturn<ExtArgs> | null
}

/**
 * ap_build_items upsert
 */
export type ap_build_itemsUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * The filter to search for the ap_build_items to update in case it exists.
   */
  where: Prisma.ap_build_itemsWhereUniqueInput
  /**
   * In case the ap_build_items found by the `where` argument doesn't exist, create a new ap_build_items with this data.
   */
  create: Prisma.XOR<Prisma.ap_build_itemsCreateInput, Prisma.ap_build_itemsUncheckedCreateInput>
  /**
   * In case the ap_build_items was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.ap_build_itemsUpdateInput, Prisma.ap_build_itemsUncheckedUpdateInput>
}

/**
 * ap_build_items delete
 */
export type ap_build_itemsDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  /**
   * Filter which ap_build_items to delete.
   */
  where: Prisma.ap_build_itemsWhereUniqueInput
}

/**
 * ap_build_items deleteMany
 */
export type ap_build_itemsDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which ap_build_items to delete
   */
  where?: Prisma.ap_build_itemsWhereInput
  /**
   * Limit how many ap_build_items to delete.
   */
  limit?: number
}

/**
 * ap_build_items without action
 */
export type ap_build_itemsDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
}
