// @prisma/client/models/champions.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `champions` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/library"
import type * as $Enums from "../enums.ts"
import type * as Prisma from "../internal/prismaNamespace.ts"

/**
 * Model champions
 * 
 */
export type championsModel = runtime.Types.Result.DefaultSelection<Prisma.$championsPayload>

export type AggregateChampions = {
  _count: ChampionsCountAggregateOutputType | null
  _avg: ChampionsAvgAggregateOutputType | null
  _sum: ChampionsSumAggregateOutputType | null
  _min: ChampionsMinAggregateOutputType | null
  _max: ChampionsMaxAggregateOutputType | null
}

export type ChampionsAvgAggregateOutputType = {
  id: number | null
  pick_rate: number | null
  win_rate: number | null
}

export type ChampionsSumAggregateOutputType = {
  id: number | null
  pick_rate: number | null
  win_rate: number | null
}

export type ChampionsMinAggregateOutputType = {
  id: number | null
  name: string | null
  title: string | null
  description: string | null
  role: string | null
  image_url: string | null
  pick_rate: number | null
  win_rate: number | null
  riot_id: string | null
  created_at: Date | null
}

export type ChampionsMaxAggregateOutputType = {
  id: number | null
  name: string | null
  title: string | null
  description: string | null
  role: string | null
  image_url: string | null
  pick_rate: number | null
  win_rate: number | null
  riot_id: string | null
  created_at: Date | null
}

export type ChampionsCountAggregateOutputType = {
  id: number
  name: number
  title: number
  description: number
  role: number
  image_url: number
  pick_rate: number
  win_rate: number
  riot_id: number
  created_at: number
  _all: number
}


export type ChampionsAvgAggregateInputType = {
  id?: true
  pick_rate?: true
  win_rate?: true
}

export type ChampionsSumAggregateInputType = {
  id?: true
  pick_rate?: true
  win_rate?: true
}

export type ChampionsMinAggregateInputType = {
  id?: true
  name?: true
  title?: true
  description?: true
  role?: true
  image_url?: true
  pick_rate?: true
  win_rate?: true
  riot_id?: true
  created_at?: true
}

export type ChampionsMaxAggregateInputType = {
  id?: true
  name?: true
  title?: true
  description?: true
  role?: true
  image_url?: true
  pick_rate?: true
  win_rate?: true
  riot_id?: true
  created_at?: true
}

export type ChampionsCountAggregateInputType = {
  id?: true
  name?: true
  title?: true
  description?: true
  role?: true
  image_url?: true
  pick_rate?: true
  win_rate?: true
  riot_id?: true
  created_at?: true
  _all?: true
}

export type ChampionsAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which champions to aggregate.
   */
  where?: Prisma.championsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of champions to fetch.
   */
  orderBy?: Prisma.championsOrderByWithRelationInput | Prisma.championsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.championsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` champions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` champions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned champions
  **/
  _count?: true | ChampionsCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to average
  **/
  _avg?: ChampionsAvgAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to sum
  **/
  _sum?: ChampionsSumAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: ChampionsMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: ChampionsMaxAggregateInputType
}

export type GetChampionsAggregateType<T extends ChampionsAggregateArgs> = {
      [P in keyof T & keyof AggregateChampions]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateChampions[P]>
    : Prisma.GetScalarType<T[P], AggregateChampions[P]>
}




export type championsGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.championsWhereInput
  orderBy?: Prisma.championsOrderByWithAggregationInput | Prisma.championsOrderByWithAggregationInput[]
  by: Prisma.ChampionsScalarFieldEnum[] | Prisma.ChampionsScalarFieldEnum
  having?: Prisma.championsScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: ChampionsCountAggregateInputType | true
  _avg?: ChampionsAvgAggregateInputType
  _sum?: ChampionsSumAggregateInputType
  _min?: ChampionsMinAggregateInputType
  _max?: ChampionsMaxAggregateInputType
}

export type ChampionsGroupByOutputType = {
  id: number
  name: string
  title: string
  description: string | null
  role: string
  image_url: string | null
  pick_rate: number | null
  win_rate: number | null
  riot_id: string | null
  created_at: Date
  _count: ChampionsCountAggregateOutputType | null
  _avg: ChampionsAvgAggregateOutputType | null
  _sum: ChampionsSumAggregateOutputType | null
  _min: ChampionsMinAggregateOutputType | null
  _max: ChampionsMaxAggregateOutputType | null
}

type GetChampionsGroupByPayload<T extends championsGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<ChampionsGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof ChampionsGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], ChampionsGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], ChampionsGroupByOutputType[P]>
      }
    >
  >



export type championsWhereInput = {
  AND?: Prisma.championsWhereInput | Prisma.championsWhereInput[]
  OR?: Prisma.championsWhereInput[]
  NOT?: Prisma.championsWhereInput | Prisma.championsWhereInput[]
  id?: Prisma.IntFilter<"champions"> | number
  name?: Prisma.StringFilter<"champions"> | string
  title?: Prisma.StringFilter<"champions"> | string
  description?: Prisma.StringNullableFilter<"champions"> | string | null
  role?: Prisma.StringFilter<"champions"> | string
  image_url?: Prisma.StringNullableFilter<"champions"> | string | null
  pick_rate?: Prisma.FloatNullableFilter<"champions"> | number | null
  win_rate?: Prisma.FloatNullableFilter<"champions"> | number | null
  riot_id?: Prisma.StringNullableFilter<"champions"> | string | null
  created_at?: Prisma.DateTimeFilter<"champions"> | Date | string
  ap_builds?: Prisma.Ap_buildsListRelationFilter
  favorite_champions?: Prisma.Favorite_championsListRelationFilter
  player_statistics?: Prisma.Player_statisticsListRelationFilter
}

export type championsOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  title?: Prisma.SortOrder
  description?: Prisma.SortOrderInput | Prisma.SortOrder
  role?: Prisma.SortOrder
  image_url?: Prisma.SortOrderInput | Prisma.SortOrder
  pick_rate?: Prisma.SortOrderInput | Prisma.SortOrder
  win_rate?: Prisma.SortOrderInput | Prisma.SortOrder
  riot_id?: Prisma.SortOrderInput | Prisma.SortOrder
  created_at?: Prisma.SortOrder
  ap_builds?: Prisma.ap_buildsOrderByRelationAggregateInput
  favorite_champions?: Prisma.favorite_championsOrderByRelationAggregateInput
  player_statistics?: Prisma.player_statisticsOrderByRelationAggregateInput
}

export type championsWhereUniqueInput = Prisma.AtLeast<{
  id?: number
  name?: string
  AND?: Prisma.championsWhereInput | Prisma.championsWhereInput[]
  OR?: Prisma.championsWhereInput[]
  NOT?: Prisma.championsWhereInput | Prisma.championsWhereInput[]
  title?: Prisma.StringFilter<"champions"> | string
  description?: Prisma.StringNullableFilter<"champions"> | string | null
  role?: Prisma.StringFilter<"champions"> | string
  image_url?: Prisma.StringNullableFilter<"champions"> | string | null
  pick_rate?: Prisma.FloatNullableFilter<"champions"> | number | null
  win_rate?: Prisma.FloatNullableFilter<"champions"> | number | null
  riot_id?: Prisma.StringNullableFilter<"champions"> | string | null
  created_at?: Prisma.DateTimeFilter<"champions"> | Date | string
  ap_builds?: Prisma.Ap_buildsListRelationFilter
  favorite_champions?: Prisma.Favorite_championsListRelationFilter
  player_statistics?: Prisma.Player_statisticsListRelationFilter
}, "id" | "name">

export type championsOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  title?: Prisma.SortOrder
  description?: Prisma.SortOrderInput | Prisma.SortOrder
  role?: Prisma.SortOrder
  image_url?: Prisma.SortOrderInput | Prisma.SortOrder
  pick_rate?: Prisma.SortOrderInput | Prisma.SortOrder
  win_rate?: Prisma.SortOrderInput | Prisma.SortOrder
  riot_id?: Prisma.SortOrderInput | Prisma.SortOrder
  created_at?: Prisma.SortOrder
  _count?: Prisma.championsCountOrderByAggregateInput
  _avg?: Prisma.championsAvgOrderByAggregateInput
  _max?: Prisma.championsMaxOrderByAggregateInput
  _min?: Prisma.championsMinOrderByAggregateInput
  _sum?: Prisma.championsSumOrderByAggregateInput
}

export type championsScalarWhereWithAggregatesInput = {
  AND?: Prisma.championsScalarWhereWithAggregatesInput | Prisma.championsScalarWhereWithAggregatesInput[]
  OR?: Prisma.championsScalarWhereWithAggregatesInput[]
  NOT?: Prisma.championsScalarWhereWithAggregatesInput | Prisma.championsScalarWhereWithAggregatesInput[]
  id?: Prisma.IntWithAggregatesFilter<"champions"> | number
  name?: Prisma.StringWithAggregatesFilter<"champions"> | string
  title?: Prisma.StringWithAggregatesFilter<"champions"> | string
  description?: Prisma.StringNullableWithAggregatesFilter<"champions"> | string | null
  role?: Prisma.StringWithAggregatesFilter<"champions"> | string
  image_url?: Prisma.StringNullableWithAggregatesFilter<"champions"> | string | null
  pick_rate?: Prisma.FloatNullableWithAggregatesFilter<"champions"> | number | null
  win_rate?: Prisma.FloatNullableWithAggregatesFilter<"champions"> | number | null
  riot_id?: Prisma.StringNullableWithAggregatesFilter<"champions"> | string | null
  created_at?: Prisma.DateTimeWithAggregatesFilter<"champions"> | Date | string
}

export type championsCreateInput = {
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
  ap_builds?: Prisma.ap_buildsCreateNestedManyWithoutChampionsInput
  favorite_champions?: Prisma.favorite_championsCreateNestedManyWithoutChampionsInput
  player_statistics?: Prisma.player_statisticsCreateNestedManyWithoutChampionsInput
}

export type championsUncheckedCreateInput = {
  id?: number
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
  ap_builds?: Prisma.ap_buildsUncheckedCreateNestedManyWithoutChampionsInput
  favorite_champions?: Prisma.favorite_championsUncheckedCreateNestedManyWithoutChampionsInput
  player_statistics?: Prisma.player_statisticsUncheckedCreateNestedManyWithoutChampionsInput
}

export type championsUpdateInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  ap_builds?: Prisma.ap_buildsUpdateManyWithoutChampionsNestedInput
  favorite_champions?: Prisma.favorite_championsUpdateManyWithoutChampionsNestedInput
  player_statistics?: Prisma.player_statisticsUpdateManyWithoutChampionsNestedInput
}

export type championsUncheckedUpdateInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  ap_builds?: Prisma.ap_buildsUncheckedUpdateManyWithoutChampionsNestedInput
  favorite_champions?: Prisma.favorite_championsUncheckedUpdateManyWithoutChampionsNestedInput
  player_statistics?: Prisma.player_statisticsUncheckedUpdateManyWithoutChampionsNestedInput
}

export type championsCreateManyInput = {
  id?: number
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
}

export type championsUpdateManyMutationInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type championsUncheckedUpdateManyInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ChampionsScalarRelationFilter = {
  is?: Prisma.championsWhereInput
  isNot?: Prisma.championsWhereInput
}

export type championsCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  title?: Prisma.SortOrder
  description?: Prisma.SortOrder
  role?: Prisma.SortOrder
  image_url?: Prisma.SortOrder
  pick_rate?: Prisma.SortOrder
  win_rate?: Prisma.SortOrder
  riot_id?: Prisma.SortOrder
  created_at?: Prisma.SortOrder
}

export type championsAvgOrderByAggregateInput = {
  id?: Prisma.SortOrder
  pick_rate?: Prisma.SortOrder
  win_rate?: Prisma.SortOrder
}

export type championsMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  title?: Prisma.SortOrder
  description?: Prisma.SortOrder
  role?: Prisma.SortOrder
  image_url?: Prisma.SortOrder
  pick_rate?: Prisma.SortOrder
  win_rate?: Prisma.SortOrder
  riot_id?: Prisma.SortOrder
  created_at?: Prisma.SortOrder
}

export type championsMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  title?: Prisma.SortOrder
  description?: Prisma.SortOrder
  role?: Prisma.SortOrder
  image_url?: Prisma.SortOrder
  pick_rate?: Prisma.SortOrder
  win_rate?: Prisma.SortOrder
  riot_id?: Prisma.SortOrder
  created_at?: Prisma.SortOrder
}

export type championsSumOrderByAggregateInput = {
  id?: Prisma.SortOrder
  pick_rate?: Prisma.SortOrder
  win_rate?: Prisma.SortOrder
}

export type championsCreateNestedOneWithoutAp_buildsInput = {
  create?: Prisma.XOR<Prisma.championsCreateWithoutAp_buildsInput, Prisma.championsUncheckedCreateWithoutAp_buildsInput>
  connectOrCreate?: Prisma.championsCreateOrConnectWithoutAp_buildsInput
  connect?: Prisma.championsWhereUniqueInput
}

export type championsUpdateOneRequiredWithoutAp_buildsNestedInput = {
  create?: Prisma.XOR<Prisma.championsCreateWithoutAp_buildsInput, Prisma.championsUncheckedCreateWithoutAp_buildsInput>
  connectOrCreate?: Prisma.championsCreateOrConnectWithoutAp_buildsInput
  upsert?: Prisma.championsUpsertWithoutAp_buildsInput
  connect?: Prisma.championsWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.championsUpdateToOneWithWhereWithoutAp_buildsInput, Prisma.championsUpdateWithoutAp_buildsInput>, Prisma.championsUncheckedUpdateWithoutAp_buildsInput>
}

export type championsCreateNestedOneWithoutFavorite_championsInput = {
  create?: Prisma.XOR<Prisma.championsCreateWithoutFavorite_championsInput, Prisma.championsUncheckedCreateWithoutFavorite_championsInput>
  connectOrCreate?: Prisma.championsCreateOrConnectWithoutFavorite_championsInput
  connect?: Prisma.championsWhereUniqueInput
}

export type championsUpdateOneRequiredWithoutFavorite_championsNestedInput = {
  create?: Prisma.XOR<Prisma.championsCreateWithoutFavorite_championsInput, Prisma.championsUncheckedCreateWithoutFavorite_championsInput>
  connectOrCreate?: Prisma.championsCreateOrConnectWithoutFavorite_championsInput
  upsert?: Prisma.championsUpsertWithoutFavorite_championsInput
  connect?: Prisma.championsWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.championsUpdateToOneWithWhereWithoutFavorite_championsInput, Prisma.championsUpdateWithoutFavorite_championsInput>, Prisma.championsUncheckedUpdateWithoutFavorite_championsInput>
}

export type championsCreateNestedOneWithoutPlayer_statisticsInput = {
  create?: Prisma.XOR<Prisma.championsCreateWithoutPlayer_statisticsInput, Prisma.championsUncheckedCreateWithoutPlayer_statisticsInput>
  connectOrCreate?: Prisma.championsCreateOrConnectWithoutPlayer_statisticsInput
  connect?: Prisma.championsWhereUniqueInput
}

export type championsUpdateOneRequiredWithoutPlayer_statisticsNestedInput = {
  create?: Prisma.XOR<Prisma.championsCreateWithoutPlayer_statisticsInput, Prisma.championsUncheckedCreateWithoutPlayer_statisticsInput>
  connectOrCreate?: Prisma.championsCreateOrConnectWithoutPlayer_statisticsInput
  upsert?: Prisma.championsUpsertWithoutPlayer_statisticsInput
  connect?: Prisma.championsWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.championsUpdateToOneWithWhereWithoutPlayer_statisticsInput, Prisma.championsUpdateWithoutPlayer_statisticsInput>, Prisma.championsUncheckedUpdateWithoutPlayer_statisticsInput>
}

export type championsCreateWithoutAp_buildsInput = {
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
  favorite_champions?: Prisma.favorite_championsCreateNestedManyWithoutChampionsInput
  player_statistics?: Prisma.player_statisticsCreateNestedManyWithoutChampionsInput
}

export type championsUncheckedCreateWithoutAp_buildsInput = {
  id?: number
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
  favorite_champions?: Prisma.favorite_championsUncheckedCreateNestedManyWithoutChampionsInput
  player_statistics?: Prisma.player_statisticsUncheckedCreateNestedManyWithoutChampionsInput
}

export type championsCreateOrConnectWithoutAp_buildsInput = {
  where: Prisma.championsWhereUniqueInput
  create: Prisma.XOR<Prisma.championsCreateWithoutAp_buildsInput, Prisma.championsUncheckedCreateWithoutAp_buildsInput>
}

export type championsUpsertWithoutAp_buildsInput = {
  update: Prisma.XOR<Prisma.championsUpdateWithoutAp_buildsInput, Prisma.championsUncheckedUpdateWithoutAp_buildsInput>
  create: Prisma.XOR<Prisma.championsCreateWithoutAp_buildsInput, Prisma.championsUncheckedCreateWithoutAp_buildsInput>
  where?: Prisma.championsWhereInput
}

export type championsUpdateToOneWithWhereWithoutAp_buildsInput = {
  where?: Prisma.championsWhereInput
  data: Prisma.XOR<Prisma.championsUpdateWithoutAp_buildsInput, Prisma.championsUncheckedUpdateWithoutAp_buildsInput>
}

export type championsUpdateWithoutAp_buildsInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  favorite_champions?: Prisma.favorite_championsUpdateManyWithoutChampionsNestedInput
  player_statistics?: Prisma.player_statisticsUpdateManyWithoutChampionsNestedInput
}

export type championsUncheckedUpdateWithoutAp_buildsInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  favorite_champions?: Prisma.favorite_championsUncheckedUpdateManyWithoutChampionsNestedInput
  player_statistics?: Prisma.player_statisticsUncheckedUpdateManyWithoutChampionsNestedInput
}

export type championsCreateWithoutFavorite_championsInput = {
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
  ap_builds?: Prisma.ap_buildsCreateNestedManyWithoutChampionsInput
  player_statistics?: Prisma.player_statisticsCreateNestedManyWithoutChampionsInput
}

export type championsUncheckedCreateWithoutFavorite_championsInput = {
  id?: number
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
  ap_builds?: Prisma.ap_buildsUncheckedCreateNestedManyWithoutChampionsInput
  player_statistics?: Prisma.player_statisticsUncheckedCreateNestedManyWithoutChampionsInput
}

export type championsCreateOrConnectWithoutFavorite_championsInput = {
  where: Prisma.championsWhereUniqueInput
  create: Prisma.XOR<Prisma.championsCreateWithoutFavorite_championsInput, Prisma.championsUncheckedCreateWithoutFavorite_championsInput>
}

export type championsUpsertWithoutFavorite_championsInput = {
  update: Prisma.XOR<Prisma.championsUpdateWithoutFavorite_championsInput, Prisma.championsUncheckedUpdateWithoutFavorite_championsInput>
  create: Prisma.XOR<Prisma.championsCreateWithoutFavorite_championsInput, Prisma.championsUncheckedCreateWithoutFavorite_championsInput>
  where?: Prisma.championsWhereInput
}

export type championsUpdateToOneWithWhereWithoutFavorite_championsInput = {
  where?: Prisma.championsWhereInput
  data: Prisma.XOR<Prisma.championsUpdateWithoutFavorite_championsInput, Prisma.championsUncheckedUpdateWithoutFavorite_championsInput>
}

export type championsUpdateWithoutFavorite_championsInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  ap_builds?: Prisma.ap_buildsUpdateManyWithoutChampionsNestedInput
  player_statistics?: Prisma.player_statisticsUpdateManyWithoutChampionsNestedInput
}

export type championsUncheckedUpdateWithoutFavorite_championsInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  ap_builds?: Prisma.ap_buildsUncheckedUpdateManyWithoutChampionsNestedInput
  player_statistics?: Prisma.player_statisticsUncheckedUpdateManyWithoutChampionsNestedInput
}

export type championsCreateWithoutPlayer_statisticsInput = {
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
  ap_builds?: Prisma.ap_buildsCreateNestedManyWithoutChampionsInput
  favorite_champions?: Prisma.favorite_championsCreateNestedManyWithoutChampionsInput
}

export type championsUncheckedCreateWithoutPlayer_statisticsInput = {
  id?: number
  name: string
  title: string
  description?: string | null
  role: string
  image_url?: string | null
  pick_rate?: number | null
  win_rate?: number | null
  riot_id?: string | null
  created_at: Date | string
  ap_builds?: Prisma.ap_buildsUncheckedCreateNestedManyWithoutChampionsInput
  favorite_champions?: Prisma.favorite_championsUncheckedCreateNestedManyWithoutChampionsInput
}

export type championsCreateOrConnectWithoutPlayer_statisticsInput = {
  where: Prisma.championsWhereUniqueInput
  create: Prisma.XOR<Prisma.championsCreateWithoutPlayer_statisticsInput, Prisma.championsUncheckedCreateWithoutPlayer_statisticsInput>
}

export type championsUpsertWithoutPlayer_statisticsInput = {
  update: Prisma.XOR<Prisma.championsUpdateWithoutPlayer_statisticsInput, Prisma.championsUncheckedUpdateWithoutPlayer_statisticsInput>
  create: Prisma.XOR<Prisma.championsCreateWithoutPlayer_statisticsInput, Prisma.championsUncheckedCreateWithoutPlayer_statisticsInput>
  where?: Prisma.championsWhereInput
}

export type championsUpdateToOneWithWhereWithoutPlayer_statisticsInput = {
  where?: Prisma.championsWhereInput
  data: Prisma.XOR<Prisma.championsUpdateWithoutPlayer_statisticsInput, Prisma.championsUncheckedUpdateWithoutPlayer_statisticsInput>
}

export type championsUpdateWithoutPlayer_statisticsInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  ap_builds?: Prisma.ap_buildsUpdateManyWithoutChampionsNestedInput
  favorite_champions?: Prisma.favorite_championsUpdateManyWithoutChampionsNestedInput
}

export type championsUncheckedUpdateWithoutPlayer_statisticsInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  name?: Prisma.StringFieldUpdateOperationsInput | string
  title?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  role?: Prisma.StringFieldUpdateOperationsInput | string
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  pick_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  win_rate?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  ap_builds?: Prisma.ap_buildsUncheckedUpdateManyWithoutChampionsNestedInput
  favorite_champions?: Prisma.favorite_championsUncheckedUpdateManyWithoutChampionsNestedInput
}


/**
 * Count Type ChampionsCountOutputType
 */

export type ChampionsCountOutputType = {
  ap_builds: number
  favorite_champions: number
  player_statistics: number
}

export type ChampionsCountOutputTypeSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  ap_builds?: boolean | ChampionsCountOutputTypeCountAp_buildsArgs
  favorite_champions?: boolean | ChampionsCountOutputTypeCountFavorite_championsArgs
  player_statistics?: boolean | ChampionsCountOutputTypeCountPlayer_statisticsArgs
}

/**
 * ChampionsCountOutputType without action
 */
export type ChampionsCountOutputTypeDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ChampionsCountOutputType
   */
  select?: Prisma.ChampionsCountOutputTypeSelect<ExtArgs> | null
}

/**
 * ChampionsCountOutputType without action
 */
export type ChampionsCountOutputTypeCountAp_buildsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.ap_buildsWhereInput
}

/**
 * ChampionsCountOutputType without action
 */
export type ChampionsCountOutputTypeCountFavorite_championsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.favorite_championsWhereInput
}

/**
 * ChampionsCountOutputType without action
 */
export type ChampionsCountOutputTypeCountPlayer_statisticsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.player_statisticsWhereInput
}


export type championsSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  name?: boolean
  title?: boolean
  description?: boolean
  role?: boolean
  image_url?: boolean
  pick_rate?: boolean
  win_rate?: boolean
  riot_id?: boolean
  created_at?: boolean
  ap_builds?: boolean | Prisma.champions$ap_buildsArgs<ExtArgs>
  favorite_champions?: boolean | Prisma.champions$favorite_championsArgs<ExtArgs>
  player_statistics?: boolean | Prisma.champions$player_statisticsArgs<ExtArgs>
  _count?: boolean | Prisma.ChampionsCountOutputTypeDefaultArgs<ExtArgs>
}, ExtArgs["result"]["champions"]>

export type championsSelectCreateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  name?: boolean
  title?: boolean
  description?: boolean
  role?: boolean
  image_url?: boolean
  pick_rate?: boolean
  win_rate?: boolean
  riot_id?: boolean
  created_at?: boolean
}, ExtArgs["result"]["champions"]>

export type championsSelectUpdateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  name?: boolean
  title?: boolean
  description?: boolean
  role?: boolean
  image_url?: boolean
  pick_rate?: boolean
  win_rate?: boolean
  riot_id?: boolean
  created_at?: boolean
}, ExtArgs["result"]["champions"]>

export type championsSelectScalar = {
  id?: boolean
  name?: boolean
  title?: boolean
  description?: boolean
  role?: boolean
  image_url?: boolean
  pick_rate?: boolean
  win_rate?: boolean
  riot_id?: boolean
  created_at?: boolean
}

export type championsOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "name" | "title" | "description" | "role" | "image_url" | "pick_rate" | "win_rate" | "riot_id" | "created_at", ExtArgs["result"]["champions"]>
export type championsInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  ap_builds?: boolean | Prisma.champions$ap_buildsArgs<ExtArgs>
  favorite_champions?: boolean | Prisma.champions$favorite_championsArgs<ExtArgs>
  player_statistics?: boolean | Prisma.champions$player_statisticsArgs<ExtArgs>
  _count?: boolean | Prisma.ChampionsCountOutputTypeDefaultArgs<ExtArgs>
}
export type championsIncludeCreateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {}
export type championsIncludeUpdateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {}

export type $championsPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "champions"
  objects: {
    ap_builds: Prisma.$ap_buildsPayload<ExtArgs>[]
    favorite_champions: Prisma.$favorite_championsPayload<ExtArgs>[]
    player_statistics: Prisma.$player_statisticsPayload<ExtArgs>[]
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: number
    name: string
    title: string
    description: string | null
    role: string
    image_url: string | null
    pick_rate: number | null
    win_rate: number | null
    riot_id: string | null
    created_at: Date
  }, ExtArgs["result"]["champions"]>
  composites: {}
}

export type championsGetPayload<S extends boolean | null | undefined | championsDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$championsPayload, S>

export type championsCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<championsFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: ChampionsCountAggregateInputType | true
  }

export interface championsDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['champions'], meta: { name: 'champions' } }
  /**
   * Find zero or one Champions that matches the filter.
   * @param {championsFindUniqueArgs} args - Arguments to find a Champions
   * @example
   * // Get one Champions
   * const champions = await prisma.champions.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends championsFindUniqueArgs>(args: Prisma.SelectSubset<T, championsFindUniqueArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one Champions that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {championsFindUniqueOrThrowArgs} args - Arguments to find a Champions
   * @example
   * // Get one Champions
   * const champions = await prisma.champions.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends championsFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, championsFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Champions that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {championsFindFirstArgs} args - Arguments to find a Champions
   * @example
   * // Get one Champions
   * const champions = await prisma.champions.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends championsFindFirstArgs>(args?: Prisma.SelectSubset<T, championsFindFirstArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Champions that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {championsFindFirstOrThrowArgs} args - Arguments to find a Champions
   * @example
   * // Get one Champions
   * const champions = await prisma.champions.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends championsFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, championsFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Champions that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {championsFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all Champions
   * const champions = await prisma.champions.findMany()
   * 
   * // Get first 10 Champions
   * const champions = await prisma.champions.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const championsWithIdOnly = await prisma.champions.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends championsFindManyArgs>(args?: Prisma.SelectSubset<T, championsFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a Champions.
   * @param {championsCreateArgs} args - Arguments to create a Champions.
   * @example
   * // Create one Champions
   * const Champions = await prisma.champions.create({
   *   data: {
   *     // ... data to create a Champions
   *   }
   * })
   * 
   */
  create<T extends championsCreateArgs>(args: Prisma.SelectSubset<T, championsCreateArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many Champions.
   * @param {championsCreateManyArgs} args - Arguments to create many Champions.
   * @example
   * // Create many Champions
   * const champions = await prisma.champions.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends championsCreateManyArgs>(args?: Prisma.SelectSubset<T, championsCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create many Champions and returns the data saved in the database.
   * @param {championsCreateManyAndReturnArgs} args - Arguments to create many Champions.
   * @example
   * // Create many Champions
   * const champions = await prisma.champions.createManyAndReturn({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * 
   * // Create many Champions and only return the `id`
   * const championsWithIdOnly = await prisma.champions.createManyAndReturn({
   *   select: { id: true },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * 
   */
  createManyAndReturn<T extends championsCreateManyAndReturnArgs>(args?: Prisma.SelectSubset<T, championsCreateManyAndReturnArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "createManyAndReturn", GlobalOmitOptions>>

  /**
   * Delete a Champions.
   * @param {championsDeleteArgs} args - Arguments to delete one Champions.
   * @example
   * // Delete one Champions
   * const Champions = await prisma.champions.delete({
   *   where: {
   *     // ... filter to delete one Champions
   *   }
   * })
   * 
   */
  delete<T extends championsDeleteArgs>(args: Prisma.SelectSubset<T, championsDeleteArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one Champions.
   * @param {championsUpdateArgs} args - Arguments to update one Champions.
   * @example
   * // Update one Champions
   * const champions = await prisma.champions.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends championsUpdateArgs>(args: Prisma.SelectSubset<T, championsUpdateArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more Champions.
   * @param {championsDeleteManyArgs} args - Arguments to filter Champions to delete.
   * @example
   * // Delete a few Champions
   * const { count } = await prisma.champions.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends championsDeleteManyArgs>(args?: Prisma.SelectSubset<T, championsDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Champions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {championsUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many Champions
   * const champions = await prisma.champions.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends championsUpdateManyArgs>(args: Prisma.SelectSubset<T, championsUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Champions and returns the data updated in the database.
   * @param {championsUpdateManyAndReturnArgs} args - Arguments to update many Champions.
   * @example
   * // Update many Champions
   * const champions = await prisma.champions.updateManyAndReturn({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * 
   * // Update zero or more Champions and only return the `id`
   * const championsWithIdOnly = await prisma.champions.updateManyAndReturn({
   *   select: { id: true },
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * 
   */
  updateManyAndReturn<T extends championsUpdateManyAndReturnArgs>(args: Prisma.SelectSubset<T, championsUpdateManyAndReturnArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "updateManyAndReturn", GlobalOmitOptions>>

  /**
   * Create or update one Champions.
   * @param {championsUpsertArgs} args - Arguments to update or create a Champions.
   * @example
   * // Update or create a Champions
   * const champions = await prisma.champions.upsert({
   *   create: {
   *     // ... data to create a Champions
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the Champions we want to update
   *   }
   * })
   */
  upsert<T extends championsUpsertArgs>(args: Prisma.SelectSubset<T, championsUpsertArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>


  /**
   * Count the number of Champions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {championsCountArgs} args - Arguments to filter Champions to count.
   * @example
   * // Count the number of Champions
   * const count = await prisma.champions.count({
   *   where: {
   *     // ... the filter for the Champions we want to count
   *   }
   * })
  **/
  count<T extends championsCountArgs>(
    args?: Prisma.Subset<T, championsCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], ChampionsCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a Champions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ChampionsAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends ChampionsAggregateArgs>(args: Prisma.Subset<T, ChampionsAggregateArgs>): Prisma.PrismaPromise<GetChampionsAggregateType<T>>

  /**
   * Group by Champions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {championsGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends championsGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: championsGroupByArgs['orderBy'] }
      : { orderBy?: championsGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, championsGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetChampionsGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the champions model
 */
readonly fields: championsFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for champions.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__championsClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  ap_builds<T extends Prisma.champions$ap_buildsArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.champions$ap_buildsArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ap_buildsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  favorite_champions<T extends Prisma.champions$favorite_championsArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.champions$favorite_championsArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  player_statistics<T extends Prisma.champions$player_statisticsArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.champions$player_statisticsArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$player_statisticsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the champions model
 */
export interface championsFieldRefs {
  readonly id: Prisma.FieldRef<"champions", 'Int'>
  readonly name: Prisma.FieldRef<"champions", 'String'>
  readonly title: Prisma.FieldRef<"champions", 'String'>
  readonly description: Prisma.FieldRef<"champions", 'String'>
  readonly role: Prisma.FieldRef<"champions", 'String'>
  readonly image_url: Prisma.FieldRef<"champions", 'String'>
  readonly pick_rate: Prisma.FieldRef<"champions", 'Float'>
  readonly win_rate: Prisma.FieldRef<"champions", 'Float'>
  readonly riot_id: Prisma.FieldRef<"champions", 'String'>
  readonly created_at: Prisma.FieldRef<"champions", 'DateTime'>
}
    

// Custom InputTypes
/**
 * champions findUnique
 */
export type championsFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * Filter, which champions to fetch.
   */
  where: Prisma.championsWhereUniqueInput
}

/**
 * champions findUniqueOrThrow
 */
export type championsFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * Filter, which champions to fetch.
   */
  where: Prisma.championsWhereUniqueInput
}

/**
 * champions findFirst
 */
export type championsFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * Filter, which champions to fetch.
   */
  where?: Prisma.championsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of champions to fetch.
   */
  orderBy?: Prisma.championsOrderByWithRelationInput | Prisma.championsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for champions.
   */
  cursor?: Prisma.championsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` champions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` champions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of champions.
   */
  distinct?: Prisma.ChampionsScalarFieldEnum | Prisma.ChampionsScalarFieldEnum[]
}

/**
 * champions findFirstOrThrow
 */
export type championsFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * Filter, which champions to fetch.
   */
  where?: Prisma.championsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of champions to fetch.
   */
  orderBy?: Prisma.championsOrderByWithRelationInput | Prisma.championsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for champions.
   */
  cursor?: Prisma.championsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` champions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` champions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of champions.
   */
  distinct?: Prisma.ChampionsScalarFieldEnum | Prisma.ChampionsScalarFieldEnum[]
}

/**
 * champions findMany
 */
export type championsFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * Filter, which champions to fetch.
   */
  where?: Prisma.championsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of champions to fetch.
   */
  orderBy?: Prisma.championsOrderByWithRelationInput | Prisma.championsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing champions.
   */
  cursor?: Prisma.championsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` champions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` champions.
   */
  skip?: number
  distinct?: Prisma.ChampionsScalarFieldEnum | Prisma.ChampionsScalarFieldEnum[]
}

/**
 * champions create
 */
export type championsCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * The data needed to create a champions.
   */
  data: Prisma.XOR<Prisma.championsCreateInput, Prisma.championsUncheckedCreateInput>
}

/**
 * champions createMany
 */
export type championsCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many champions.
   */
  data: Prisma.championsCreateManyInput | Prisma.championsCreateManyInput[]
  skipDuplicates?: boolean
}

/**
 * champions createManyAndReturn
 */
export type championsCreateManyAndReturnArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelectCreateManyAndReturn<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * The data used to create many champions.
   */
  data: Prisma.championsCreateManyInput | Prisma.championsCreateManyInput[]
  skipDuplicates?: boolean
}

/**
 * champions update
 */
export type championsUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * The data needed to update a champions.
   */
  data: Prisma.XOR<Prisma.championsUpdateInput, Prisma.championsUncheckedUpdateInput>
  /**
   * Choose, which champions to update.
   */
  where: Prisma.championsWhereUniqueInput
}

/**
 * champions updateMany
 */
export type championsUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update champions.
   */
  data: Prisma.XOR<Prisma.championsUpdateManyMutationInput, Prisma.championsUncheckedUpdateManyInput>
  /**
   * Filter which champions to update
   */
  where?: Prisma.championsWhereInput
  /**
   * Limit how many champions to update.
   */
  limit?: number
}

/**
 * champions updateManyAndReturn
 */
export type championsUpdateManyAndReturnArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelectUpdateManyAndReturn<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * The data used to update champions.
   */
  data: Prisma.XOR<Prisma.championsUpdateManyMutationInput, Prisma.championsUncheckedUpdateManyInput>
  /**
   * Filter which champions to update
   */
  where?: Prisma.championsWhereInput
  /**
   * Limit how many champions to update.
   */
  limit?: number
}

/**
 * champions upsert
 */
export type championsUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * The filter to search for the champions to update in case it exists.
   */
  where: Prisma.championsWhereUniqueInput
  /**
   * In case the champions found by the `where` argument doesn't exist, create a new champions with this data.
   */
  create: Prisma.XOR<Prisma.championsCreateInput, Prisma.championsUncheckedCreateInput>
  /**
   * In case the champions was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.championsUpdateInput, Prisma.championsUncheckedUpdateInput>
}

/**
 * champions delete
 */
export type championsDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
  /**
   * Filter which champions to delete.
   */
  where: Prisma.championsWhereUniqueInput
}

/**
 * champions deleteMany
 */
export type championsDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which champions to delete
   */
  where?: Prisma.championsWhereInput
  /**
   * Limit how many champions to delete.
   */
  limit?: number
}

/**
 * champions.ap_builds
 */
export type champions$ap_buildsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_builds
   */
  select?: Prisma.ap_buildsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_builds
   */
  omit?: Prisma.ap_buildsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_buildsInclude<ExtArgs> | null
  where?: Prisma.ap_buildsWhereInput
  orderBy?: Prisma.ap_buildsOrderByWithRelationInput | Prisma.ap_buildsOrderByWithRelationInput[]
  cursor?: Prisma.ap_buildsWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.Ap_buildsScalarFieldEnum | Prisma.Ap_buildsScalarFieldEnum[]
}

/**
 * champions.favorite_champions
 */
export type champions$favorite_championsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  where?: Prisma.favorite_championsWhereInput
  orderBy?: Prisma.favorite_championsOrderByWithRelationInput | Prisma.favorite_championsOrderByWithRelationInput[]
  cursor?: Prisma.favorite_championsWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.Favorite_championsScalarFieldEnum | Prisma.Favorite_championsScalarFieldEnum[]
}

/**
 * champions.player_statistics
 */
export type champions$player_statisticsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the player_statistics
   */
  select?: Prisma.player_statisticsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the player_statistics
   */
  omit?: Prisma.player_statisticsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.player_statisticsInclude<ExtArgs> | null
  where?: Prisma.player_statisticsWhereInput
  orderBy?: Prisma.player_statisticsOrderByWithRelationInput | Prisma.player_statisticsOrderByWithRelationInput[]
  cursor?: Prisma.player_statisticsWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.Player_statisticsScalarFieldEnum | Prisma.Player_statisticsScalarFieldEnum[]
}

/**
 * champions without action
 */
export type championsDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the champions
   */
  select?: Prisma.championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the champions
   */
  omit?: Prisma.championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.championsInclude<ExtArgs> | null
}
