// @prisma/client/models/favorite_champions.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `favorite_champions` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/library"
import type * as $Enums from "../enums.ts"
import type * as Prisma from "../internal/prismaNamespace.ts"

/**
 * Model favorite_champions
 * 
 */
export type favorite_championsModel = runtime.Types.Result.DefaultSelection<Prisma.$favorite_championsPayload>

export type AggregateFavorite_champions = {
  _count: Favorite_championsCountAggregateOutputType | null
  _avg: Favorite_championsAvgAggregateOutputType | null
  _sum: Favorite_championsSumAggregateOutputType | null
  _min: Favorite_championsMinAggregateOutputType | null
  _max: Favorite_championsMaxAggregateOutputType | null
}

export type Favorite_championsAvgAggregateOutputType = {
  id: number | null
  player_id: number | null
  champion_id: number | null
}

export type Favorite_championsSumAggregateOutputType = {
  id: number | null
  player_id: number | null
  champion_id: number | null
}

export type Favorite_championsMinAggregateOutputType = {
  id: number | null
  player_id: number | null
  champion_id: number | null
  added_at: Date | null
}

export type Favorite_championsMaxAggregateOutputType = {
  id: number | null
  player_id: number | null
  champion_id: number | null
  added_at: Date | null
}

export type Favorite_championsCountAggregateOutputType = {
  id: number
  player_id: number
  champion_id: number
  added_at: number
  _all: number
}


export type Favorite_championsAvgAggregateInputType = {
  id?: true
  player_id?: true
  champion_id?: true
}

export type Favorite_championsSumAggregateInputType = {
  id?: true
  player_id?: true
  champion_id?: true
}

export type Favorite_championsMinAggregateInputType = {
  id?: true
  player_id?: true
  champion_id?: true
  added_at?: true
}

export type Favorite_championsMaxAggregateInputType = {
  id?: true
  player_id?: true
  champion_id?: true
  added_at?: true
}

export type Favorite_championsCountAggregateInputType = {
  id?: true
  player_id?: true
  champion_id?: true
  added_at?: true
  _all?: true
}

export type Favorite_championsAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which favorite_champions to aggregate.
   */
  where?: Prisma.favorite_championsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of favorite_champions to fetch.
   */
  orderBy?: Prisma.favorite_championsOrderByWithRelationInput | Prisma.favorite_championsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.favorite_championsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` favorite_champions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` favorite_champions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned favorite_champions
  **/
  _count?: true | Favorite_championsCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to average
  **/
  _avg?: Favorite_championsAvgAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to sum
  **/
  _sum?: Favorite_championsSumAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: Favorite_championsMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: Favorite_championsMaxAggregateInputType
}

export type GetFavorite_championsAggregateType<T extends Favorite_championsAggregateArgs> = {
      [P in keyof T & keyof AggregateFavorite_champions]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateFavorite_champions[P]>
    : Prisma.GetScalarType<T[P], AggregateFavorite_champions[P]>
}




export type favorite_championsGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.favorite_championsWhereInput
  orderBy?: Prisma.favorite_championsOrderByWithAggregationInput | Prisma.favorite_championsOrderByWithAggregationInput[]
  by: Prisma.Favorite_championsScalarFieldEnum[] | Prisma.Favorite_championsScalarFieldEnum
  having?: Prisma.favorite_championsScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: Favorite_championsCountAggregateInputType | true
  _avg?: Favorite_championsAvgAggregateInputType
  _sum?: Favorite_championsSumAggregateInputType
  _min?: Favorite_championsMinAggregateInputType
  _max?: Favorite_championsMaxAggregateInputType
}

export type Favorite_championsGroupByOutputType = {
  id: number
  player_id: number
  champion_id: number
  added_at: Date
  _count: Favorite_championsCountAggregateOutputType | null
  _avg: Favorite_championsAvgAggregateOutputType | null
  _sum: Favorite_championsSumAggregateOutputType | null
  _min: Favorite_championsMinAggregateOutputType | null
  _max: Favorite_championsMaxAggregateOutputType | null
}

type GetFavorite_championsGroupByPayload<T extends favorite_championsGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<Favorite_championsGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof Favorite_championsGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], Favorite_championsGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], Favorite_championsGroupByOutputType[P]>
      }
    >
  >



export type favorite_championsWhereInput = {
  AND?: Prisma.favorite_championsWhereInput | Prisma.favorite_championsWhereInput[]
  OR?: Prisma.favorite_championsWhereInput[]
  NOT?: Prisma.favorite_championsWhereInput | Prisma.favorite_championsWhereInput[]
  id?: Prisma.IntFilter<"favorite_champions"> | number
  player_id?: Prisma.IntFilter<"favorite_champions"> | number
  champion_id?: Prisma.IntFilter<"favorite_champions"> | number
  added_at?: Prisma.DateTimeFilter<"favorite_champions"> | Date | string
  champions?: Prisma.XOR<Prisma.ChampionsScalarRelationFilter, Prisma.championsWhereInput>
  players?: Prisma.XOR<Prisma.PlayersScalarRelationFilter, Prisma.playersWhereInput>
}

export type favorite_championsOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  player_id?: Prisma.SortOrder
  champion_id?: Prisma.SortOrder
  added_at?: Prisma.SortOrder
  champions?: Prisma.championsOrderByWithRelationInput
  players?: Prisma.playersOrderByWithRelationInput
}

export type favorite_championsWhereUniqueInput = Prisma.AtLeast<{
  id?: number
  player_id_champion_id?: Prisma.favorite_championsPlayer_idChampion_idCompoundUniqueInput
  AND?: Prisma.favorite_championsWhereInput | Prisma.favorite_championsWhereInput[]
  OR?: Prisma.favorite_championsWhereInput[]
  NOT?: Prisma.favorite_championsWhereInput | Prisma.favorite_championsWhereInput[]
  player_id?: Prisma.IntFilter<"favorite_champions"> | number
  champion_id?: Prisma.IntFilter<"favorite_champions"> | number
  added_at?: Prisma.DateTimeFilter<"favorite_champions"> | Date | string
  champions?: Prisma.XOR<Prisma.ChampionsScalarRelationFilter, Prisma.championsWhereInput>
  players?: Prisma.XOR<Prisma.PlayersScalarRelationFilter, Prisma.playersWhereInput>
}, "id" | "player_id_champion_id">

export type favorite_championsOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  player_id?: Prisma.SortOrder
  champion_id?: Prisma.SortOrder
  added_at?: Prisma.SortOrder
  _count?: Prisma.favorite_championsCountOrderByAggregateInput
  _avg?: Prisma.favorite_championsAvgOrderByAggregateInput
  _max?: Prisma.favorite_championsMaxOrderByAggregateInput
  _min?: Prisma.favorite_championsMinOrderByAggregateInput
  _sum?: Prisma.favorite_championsSumOrderByAggregateInput
}

export type favorite_championsScalarWhereWithAggregatesInput = {
  AND?: Prisma.favorite_championsScalarWhereWithAggregatesInput | Prisma.favorite_championsScalarWhereWithAggregatesInput[]
  OR?: Prisma.favorite_championsScalarWhereWithAggregatesInput[]
  NOT?: Prisma.favorite_championsScalarWhereWithAggregatesInput | Prisma.favorite_championsScalarWhereWithAggregatesInput[]
  id?: Prisma.IntWithAggregatesFilter<"favorite_champions"> | number
  player_id?: Prisma.IntWithAggregatesFilter<"favorite_champions"> | number
  champion_id?: Prisma.IntWithAggregatesFilter<"favorite_champions"> | number
  added_at?: Prisma.DateTimeWithAggregatesFilter<"favorite_champions"> | Date | string
}

export type favorite_championsCreateInput = {
  added_at: Date | string
  champions: Prisma.championsCreateNestedOneWithoutFavorite_championsInput
  players: Prisma.playersCreateNestedOneWithoutFavorite_championsInput
}

export type favorite_championsUncheckedCreateInput = {
  id?: number
  player_id: number
  champion_id: number
  added_at: Date | string
}

export type favorite_championsUpdateInput = {
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  champions?: Prisma.championsUpdateOneRequiredWithoutFavorite_championsNestedInput
  players?: Prisma.playersUpdateOneRequiredWithoutFavorite_championsNestedInput
}

export type favorite_championsUncheckedUpdateInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  player_id?: Prisma.IntFieldUpdateOperationsInput | number
  champion_id?: Prisma.IntFieldUpdateOperationsInput | number
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type favorite_championsCreateManyInput = {
  id?: number
  player_id: number
  champion_id: number
  added_at: Date | string
}

export type favorite_championsUpdateManyMutationInput = {
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type favorite_championsUncheckedUpdateManyInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  player_id?: Prisma.IntFieldUpdateOperationsInput | number
  champion_id?: Prisma.IntFieldUpdateOperationsInput | number
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type Favorite_championsListRelationFilter = {
  every?: Prisma.favorite_championsWhereInput
  some?: Prisma.favorite_championsWhereInput
  none?: Prisma.favorite_championsWhereInput
}

export type favorite_championsOrderByRelationAggregateInput = {
  _count?: Prisma.SortOrder
}

export type favorite_championsPlayer_idChampion_idCompoundUniqueInput = {
  player_id: number
  champion_id: number
}

export type favorite_championsCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  player_id?: Prisma.SortOrder
  champion_id?: Prisma.SortOrder
  added_at?: Prisma.SortOrder
}

export type favorite_championsAvgOrderByAggregateInput = {
  id?: Prisma.SortOrder
  player_id?: Prisma.SortOrder
  champion_id?: Prisma.SortOrder
}

export type favorite_championsMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  player_id?: Prisma.SortOrder
  champion_id?: Prisma.SortOrder
  added_at?: Prisma.SortOrder
}

export type favorite_championsMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  player_id?: Prisma.SortOrder
  champion_id?: Prisma.SortOrder
  added_at?: Prisma.SortOrder
}

export type favorite_championsSumOrderByAggregateInput = {
  id?: Prisma.SortOrder
  player_id?: Prisma.SortOrder
  champion_id?: Prisma.SortOrder
}

export type favorite_championsCreateNestedManyWithoutChampionsInput = {
  create?: Prisma.XOR<Prisma.favorite_championsCreateWithoutChampionsInput, Prisma.favorite_championsUncheckedCreateWithoutChampionsInput> | Prisma.favorite_championsCreateWithoutChampionsInput[] | Prisma.favorite_championsUncheckedCreateWithoutChampionsInput[]
  connectOrCreate?: Prisma.favorite_championsCreateOrConnectWithoutChampionsInput | Prisma.favorite_championsCreateOrConnectWithoutChampionsInput[]
  createMany?: Prisma.favorite_championsCreateManyChampionsInputEnvelope
  connect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
}

export type favorite_championsUncheckedCreateNestedManyWithoutChampionsInput = {
  create?: Prisma.XOR<Prisma.favorite_championsCreateWithoutChampionsInput, Prisma.favorite_championsUncheckedCreateWithoutChampionsInput> | Prisma.favorite_championsCreateWithoutChampionsInput[] | Prisma.favorite_championsUncheckedCreateWithoutChampionsInput[]
  connectOrCreate?: Prisma.favorite_championsCreateOrConnectWithoutChampionsInput | Prisma.favorite_championsCreateOrConnectWithoutChampionsInput[]
  createMany?: Prisma.favorite_championsCreateManyChampionsInputEnvelope
  connect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
}

export type favorite_championsUpdateManyWithoutChampionsNestedInput = {
  create?: Prisma.XOR<Prisma.favorite_championsCreateWithoutChampionsInput, Prisma.favorite_championsUncheckedCreateWithoutChampionsInput> | Prisma.favorite_championsCreateWithoutChampionsInput[] | Prisma.favorite_championsUncheckedCreateWithoutChampionsInput[]
  connectOrCreate?: Prisma.favorite_championsCreateOrConnectWithoutChampionsInput | Prisma.favorite_championsCreateOrConnectWithoutChampionsInput[]
  upsert?: Prisma.favorite_championsUpsertWithWhereUniqueWithoutChampionsInput | Prisma.favorite_championsUpsertWithWhereUniqueWithoutChampionsInput[]
  createMany?: Prisma.favorite_championsCreateManyChampionsInputEnvelope
  set?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  disconnect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  delete?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  connect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  update?: Prisma.favorite_championsUpdateWithWhereUniqueWithoutChampionsInput | Prisma.favorite_championsUpdateWithWhereUniqueWithoutChampionsInput[]
  updateMany?: Prisma.favorite_championsUpdateManyWithWhereWithoutChampionsInput | Prisma.favorite_championsUpdateManyWithWhereWithoutChampionsInput[]
  deleteMany?: Prisma.favorite_championsScalarWhereInput | Prisma.favorite_championsScalarWhereInput[]
}

export type favorite_championsUncheckedUpdateManyWithoutChampionsNestedInput = {
  create?: Prisma.XOR<Prisma.favorite_championsCreateWithoutChampionsInput, Prisma.favorite_championsUncheckedCreateWithoutChampionsInput> | Prisma.favorite_championsCreateWithoutChampionsInput[] | Prisma.favorite_championsUncheckedCreateWithoutChampionsInput[]
  connectOrCreate?: Prisma.favorite_championsCreateOrConnectWithoutChampionsInput | Prisma.favorite_championsCreateOrConnectWithoutChampionsInput[]
  upsert?: Prisma.favorite_championsUpsertWithWhereUniqueWithoutChampionsInput | Prisma.favorite_championsUpsertWithWhereUniqueWithoutChampionsInput[]
  createMany?: Prisma.favorite_championsCreateManyChampionsInputEnvelope
  set?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  disconnect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  delete?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  connect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  update?: Prisma.favorite_championsUpdateWithWhereUniqueWithoutChampionsInput | Prisma.favorite_championsUpdateWithWhereUniqueWithoutChampionsInput[]
  updateMany?: Prisma.favorite_championsUpdateManyWithWhereWithoutChampionsInput | Prisma.favorite_championsUpdateManyWithWhereWithoutChampionsInput[]
  deleteMany?: Prisma.favorite_championsScalarWhereInput | Prisma.favorite_championsScalarWhereInput[]
}

export type favorite_championsCreateNestedManyWithoutPlayersInput = {
  create?: Prisma.XOR<Prisma.favorite_championsCreateWithoutPlayersInput, Prisma.favorite_championsUncheckedCreateWithoutPlayersInput> | Prisma.favorite_championsCreateWithoutPlayersInput[] | Prisma.favorite_championsUncheckedCreateWithoutPlayersInput[]
  connectOrCreate?: Prisma.favorite_championsCreateOrConnectWithoutPlayersInput | Prisma.favorite_championsCreateOrConnectWithoutPlayersInput[]
  createMany?: Prisma.favorite_championsCreateManyPlayersInputEnvelope
  connect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
}

export type favorite_championsUncheckedCreateNestedManyWithoutPlayersInput = {
  create?: Prisma.XOR<Prisma.favorite_championsCreateWithoutPlayersInput, Prisma.favorite_championsUncheckedCreateWithoutPlayersInput> | Prisma.favorite_championsCreateWithoutPlayersInput[] | Prisma.favorite_championsUncheckedCreateWithoutPlayersInput[]
  connectOrCreate?: Prisma.favorite_championsCreateOrConnectWithoutPlayersInput | Prisma.favorite_championsCreateOrConnectWithoutPlayersInput[]
  createMany?: Prisma.favorite_championsCreateManyPlayersInputEnvelope
  connect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
}

export type favorite_championsUpdateManyWithoutPlayersNestedInput = {
  create?: Prisma.XOR<Prisma.favorite_championsCreateWithoutPlayersInput, Prisma.favorite_championsUncheckedCreateWithoutPlayersInput> | Prisma.favorite_championsCreateWithoutPlayersInput[] | Prisma.favorite_championsUncheckedCreateWithoutPlayersInput[]
  connectOrCreate?: Prisma.favorite_championsCreateOrConnectWithoutPlayersInput | Prisma.favorite_championsCreateOrConnectWithoutPlayersInput[]
  upsert?: Prisma.favorite_championsUpsertWithWhereUniqueWithoutPlayersInput | Prisma.favorite_championsUpsertWithWhereUniqueWithoutPlayersInput[]
  createMany?: Prisma.favorite_championsCreateManyPlayersInputEnvelope
  set?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  disconnect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  delete?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  connect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  update?: Prisma.favorite_championsUpdateWithWhereUniqueWithoutPlayersInput | Prisma.favorite_championsUpdateWithWhereUniqueWithoutPlayersInput[]
  updateMany?: Prisma.favorite_championsUpdateManyWithWhereWithoutPlayersInput | Prisma.favorite_championsUpdateManyWithWhereWithoutPlayersInput[]
  deleteMany?: Prisma.favorite_championsScalarWhereInput | Prisma.favorite_championsScalarWhereInput[]
}

export type favorite_championsUncheckedUpdateManyWithoutPlayersNestedInput = {
  create?: Prisma.XOR<Prisma.favorite_championsCreateWithoutPlayersInput, Prisma.favorite_championsUncheckedCreateWithoutPlayersInput> | Prisma.favorite_championsCreateWithoutPlayersInput[] | Prisma.favorite_championsUncheckedCreateWithoutPlayersInput[]
  connectOrCreate?: Prisma.favorite_championsCreateOrConnectWithoutPlayersInput | Prisma.favorite_championsCreateOrConnectWithoutPlayersInput[]
  upsert?: Prisma.favorite_championsUpsertWithWhereUniqueWithoutPlayersInput | Prisma.favorite_championsUpsertWithWhereUniqueWithoutPlayersInput[]
  createMany?: Prisma.favorite_championsCreateManyPlayersInputEnvelope
  set?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  disconnect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  delete?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  connect?: Prisma.favorite_championsWhereUniqueInput | Prisma.favorite_championsWhereUniqueInput[]
  update?: Prisma.favorite_championsUpdateWithWhereUniqueWithoutPlayersInput | Prisma.favorite_championsUpdateWithWhereUniqueWithoutPlayersInput[]
  updateMany?: Prisma.favorite_championsUpdateManyWithWhereWithoutPlayersInput | Prisma.favorite_championsUpdateManyWithWhereWithoutPlayersInput[]
  deleteMany?: Prisma.favorite_championsScalarWhereInput | Prisma.favorite_championsScalarWhereInput[]
}

export type favorite_championsCreateWithoutChampionsInput = {
  added_at: Date | string
  players: Prisma.playersCreateNestedOneWithoutFavorite_championsInput
}

export type favorite_championsUncheckedCreateWithoutChampionsInput = {
  id?: number
  player_id: number
  added_at: Date | string
}

export type favorite_championsCreateOrConnectWithoutChampionsInput = {
  where: Prisma.favorite_championsWhereUniqueInput
  create: Prisma.XOR<Prisma.favorite_championsCreateWithoutChampionsInput, Prisma.favorite_championsUncheckedCreateWithoutChampionsInput>
}

export type favorite_championsCreateManyChampionsInputEnvelope = {
  data: Prisma.favorite_championsCreateManyChampionsInput | Prisma.favorite_championsCreateManyChampionsInput[]
  skipDuplicates?: boolean
}

export type favorite_championsUpsertWithWhereUniqueWithoutChampionsInput = {
  where: Prisma.favorite_championsWhereUniqueInput
  update: Prisma.XOR<Prisma.favorite_championsUpdateWithoutChampionsInput, Prisma.favorite_championsUncheckedUpdateWithoutChampionsInput>
  create: Prisma.XOR<Prisma.favorite_championsCreateWithoutChampionsInput, Prisma.favorite_championsUncheckedCreateWithoutChampionsInput>
}

export type favorite_championsUpdateWithWhereUniqueWithoutChampionsInput = {
  where: Prisma.favorite_championsWhereUniqueInput
  data: Prisma.XOR<Prisma.favorite_championsUpdateWithoutChampionsInput, Prisma.favorite_championsUncheckedUpdateWithoutChampionsInput>
}

export type favorite_championsUpdateManyWithWhereWithoutChampionsInput = {
  where: Prisma.favorite_championsScalarWhereInput
  data: Prisma.XOR<Prisma.favorite_championsUpdateManyMutationInput, Prisma.favorite_championsUncheckedUpdateManyWithoutChampionsInput>
}

export type favorite_championsScalarWhereInput = {
  AND?: Prisma.favorite_championsScalarWhereInput | Prisma.favorite_championsScalarWhereInput[]
  OR?: Prisma.favorite_championsScalarWhereInput[]
  NOT?: Prisma.favorite_championsScalarWhereInput | Prisma.favorite_championsScalarWhereInput[]
  id?: Prisma.IntFilter<"favorite_champions"> | number
  player_id?: Prisma.IntFilter<"favorite_champions"> | number
  champion_id?: Prisma.IntFilter<"favorite_champions"> | number
  added_at?: Prisma.DateTimeFilter<"favorite_champions"> | Date | string
}

export type favorite_championsCreateWithoutPlayersInput = {
  added_at: Date | string
  champions: Prisma.championsCreateNestedOneWithoutFavorite_championsInput
}

export type favorite_championsUncheckedCreateWithoutPlayersInput = {
  id?: number
  champion_id: number
  added_at: Date | string
}

export type favorite_championsCreateOrConnectWithoutPlayersInput = {
  where: Prisma.favorite_championsWhereUniqueInput
  create: Prisma.XOR<Prisma.favorite_championsCreateWithoutPlayersInput, Prisma.favorite_championsUncheckedCreateWithoutPlayersInput>
}

export type favorite_championsCreateManyPlayersInputEnvelope = {
  data: Prisma.favorite_championsCreateManyPlayersInput | Prisma.favorite_championsCreateManyPlayersInput[]
  skipDuplicates?: boolean
}

export type favorite_championsUpsertWithWhereUniqueWithoutPlayersInput = {
  where: Prisma.favorite_championsWhereUniqueInput
  update: Prisma.XOR<Prisma.favorite_championsUpdateWithoutPlayersInput, Prisma.favorite_championsUncheckedUpdateWithoutPlayersInput>
  create: Prisma.XOR<Prisma.favorite_championsCreateWithoutPlayersInput, Prisma.favorite_championsUncheckedCreateWithoutPlayersInput>
}

export type favorite_championsUpdateWithWhereUniqueWithoutPlayersInput = {
  where: Prisma.favorite_championsWhereUniqueInput
  data: Prisma.XOR<Prisma.favorite_championsUpdateWithoutPlayersInput, Prisma.favorite_championsUncheckedUpdateWithoutPlayersInput>
}

export type favorite_championsUpdateManyWithWhereWithoutPlayersInput = {
  where: Prisma.favorite_championsScalarWhereInput
  data: Prisma.XOR<Prisma.favorite_championsUpdateManyMutationInput, Prisma.favorite_championsUncheckedUpdateManyWithoutPlayersInput>
}

export type favorite_championsCreateManyChampionsInput = {
  id?: number
  player_id: number
  added_at: Date | string
}

export type favorite_championsUpdateWithoutChampionsInput = {
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  players?: Prisma.playersUpdateOneRequiredWithoutFavorite_championsNestedInput
}

export type favorite_championsUncheckedUpdateWithoutChampionsInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  player_id?: Prisma.IntFieldUpdateOperationsInput | number
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type favorite_championsUncheckedUpdateManyWithoutChampionsInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  player_id?: Prisma.IntFieldUpdateOperationsInput | number
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type favorite_championsCreateManyPlayersInput = {
  id?: number
  champion_id: number
  added_at: Date | string
}

export type favorite_championsUpdateWithoutPlayersInput = {
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  champions?: Prisma.championsUpdateOneRequiredWithoutFavorite_championsNestedInput
}

export type favorite_championsUncheckedUpdateWithoutPlayersInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  champion_id?: Prisma.IntFieldUpdateOperationsInput | number
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type favorite_championsUncheckedUpdateManyWithoutPlayersInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  champion_id?: Prisma.IntFieldUpdateOperationsInput | number
  added_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}



export type favorite_championsSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  player_id?: boolean
  champion_id?: boolean
  added_at?: boolean
  champions?: boolean | Prisma.championsDefaultArgs<ExtArgs>
  players?: boolean | Prisma.playersDefaultArgs<ExtArgs>
}, ExtArgs["result"]["favorite_champions"]>

export type favorite_championsSelectCreateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  player_id?: boolean
  champion_id?: boolean
  added_at?: boolean
  champions?: boolean | Prisma.championsDefaultArgs<ExtArgs>
  players?: boolean | Prisma.playersDefaultArgs<ExtArgs>
}, ExtArgs["result"]["favorite_champions"]>

export type favorite_championsSelectUpdateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  player_id?: boolean
  champion_id?: boolean
  added_at?: boolean
  champions?: boolean | Prisma.championsDefaultArgs<ExtArgs>
  players?: boolean | Prisma.playersDefaultArgs<ExtArgs>
}, ExtArgs["result"]["favorite_champions"]>

export type favorite_championsSelectScalar = {
  id?: boolean
  player_id?: boolean
  champion_id?: boolean
  added_at?: boolean
}

export type favorite_championsOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "player_id" | "champion_id" | "added_at", ExtArgs["result"]["favorite_champions"]>
export type favorite_championsInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  champions?: boolean | Prisma.championsDefaultArgs<ExtArgs>
  players?: boolean | Prisma.playersDefaultArgs<ExtArgs>
}
export type favorite_championsIncludeCreateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  champions?: boolean | Prisma.championsDefaultArgs<ExtArgs>
  players?: boolean | Prisma.playersDefaultArgs<ExtArgs>
}
export type favorite_championsIncludeUpdateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  champions?: boolean | Prisma.championsDefaultArgs<ExtArgs>
  players?: boolean | Prisma.playersDefaultArgs<ExtArgs>
}

export type $favorite_championsPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "favorite_champions"
  objects: {
    champions: Prisma.$championsPayload<ExtArgs>
    players: Prisma.$playersPayload<ExtArgs>
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: number
    player_id: number
    champion_id: number
    added_at: Date
  }, ExtArgs["result"]["favorite_champions"]>
  composites: {}
}

export type favorite_championsGetPayload<S extends boolean | null | undefined | favorite_championsDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload, S>

export type favorite_championsCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<favorite_championsFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: Favorite_championsCountAggregateInputType | true
  }

export interface favorite_championsDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['favorite_champions'], meta: { name: 'favorite_champions' } }
  /**
   * Find zero or one Favorite_champions that matches the filter.
   * @param {favorite_championsFindUniqueArgs} args - Arguments to find a Favorite_champions
   * @example
   * // Get one Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends favorite_championsFindUniqueArgs>(args: Prisma.SelectSubset<T, favorite_championsFindUniqueArgs<ExtArgs>>): Prisma.Prisma__favorite_championsClient<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one Favorite_champions that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {favorite_championsFindUniqueOrThrowArgs} args - Arguments to find a Favorite_champions
   * @example
   * // Get one Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends favorite_championsFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, favorite_championsFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__favorite_championsClient<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Favorite_champions that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {favorite_championsFindFirstArgs} args - Arguments to find a Favorite_champions
   * @example
   * // Get one Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends favorite_championsFindFirstArgs>(args?: Prisma.SelectSubset<T, favorite_championsFindFirstArgs<ExtArgs>>): Prisma.Prisma__favorite_championsClient<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Favorite_champions that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {favorite_championsFindFirstOrThrowArgs} args - Arguments to find a Favorite_champions
   * @example
   * // Get one Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends favorite_championsFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, favorite_championsFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__favorite_championsClient<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Favorite_champions that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {favorite_championsFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.findMany()
   * 
   * // Get first 10 Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const favorite_championsWithIdOnly = await prisma.favorite_champions.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends favorite_championsFindManyArgs>(args?: Prisma.SelectSubset<T, favorite_championsFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a Favorite_champions.
   * @param {favorite_championsCreateArgs} args - Arguments to create a Favorite_champions.
   * @example
   * // Create one Favorite_champions
   * const Favorite_champions = await prisma.favorite_champions.create({
   *   data: {
   *     // ... data to create a Favorite_champions
   *   }
   * })
   * 
   */
  create<T extends favorite_championsCreateArgs>(args: Prisma.SelectSubset<T, favorite_championsCreateArgs<ExtArgs>>): Prisma.Prisma__favorite_championsClient<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many Favorite_champions.
   * @param {favorite_championsCreateManyArgs} args - Arguments to create many Favorite_champions.
   * @example
   * // Create many Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends favorite_championsCreateManyArgs>(args?: Prisma.SelectSubset<T, favorite_championsCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create many Favorite_champions and returns the data saved in the database.
   * @param {favorite_championsCreateManyAndReturnArgs} args - Arguments to create many Favorite_champions.
   * @example
   * // Create many Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.createManyAndReturn({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * 
   * // Create many Favorite_champions and only return the `id`
   * const favorite_championsWithIdOnly = await prisma.favorite_champions.createManyAndReturn({
   *   select: { id: true },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * 
   */
  createManyAndReturn<T extends favorite_championsCreateManyAndReturnArgs>(args?: Prisma.SelectSubset<T, favorite_championsCreateManyAndReturnArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "createManyAndReturn", GlobalOmitOptions>>

  /**
   * Delete a Favorite_champions.
   * @param {favorite_championsDeleteArgs} args - Arguments to delete one Favorite_champions.
   * @example
   * // Delete one Favorite_champions
   * const Favorite_champions = await prisma.favorite_champions.delete({
   *   where: {
   *     // ... filter to delete one Favorite_champions
   *   }
   * })
   * 
   */
  delete<T extends favorite_championsDeleteArgs>(args: Prisma.SelectSubset<T, favorite_championsDeleteArgs<ExtArgs>>): Prisma.Prisma__favorite_championsClient<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one Favorite_champions.
   * @param {favorite_championsUpdateArgs} args - Arguments to update one Favorite_champions.
   * @example
   * // Update one Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends favorite_championsUpdateArgs>(args: Prisma.SelectSubset<T, favorite_championsUpdateArgs<ExtArgs>>): Prisma.Prisma__favorite_championsClient<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more Favorite_champions.
   * @param {favorite_championsDeleteManyArgs} args - Arguments to filter Favorite_champions to delete.
   * @example
   * // Delete a few Favorite_champions
   * const { count } = await prisma.favorite_champions.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends favorite_championsDeleteManyArgs>(args?: Prisma.SelectSubset<T, favorite_championsDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Favorite_champions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {favorite_championsUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends favorite_championsUpdateManyArgs>(args: Prisma.SelectSubset<T, favorite_championsUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Favorite_champions and returns the data updated in the database.
   * @param {favorite_championsUpdateManyAndReturnArgs} args - Arguments to update many Favorite_champions.
   * @example
   * // Update many Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.updateManyAndReturn({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * 
   * // Update zero or more Favorite_champions and only return the `id`
   * const favorite_championsWithIdOnly = await prisma.favorite_champions.updateManyAndReturn({
   *   select: { id: true },
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * 
   */
  updateManyAndReturn<T extends favorite_championsUpdateManyAndReturnArgs>(args: Prisma.SelectSubset<T, favorite_championsUpdateManyAndReturnArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "updateManyAndReturn", GlobalOmitOptions>>

  /**
   * Create or update one Favorite_champions.
   * @param {favorite_championsUpsertArgs} args - Arguments to update or create a Favorite_champions.
   * @example
   * // Update or create a Favorite_champions
   * const favorite_champions = await prisma.favorite_champions.upsert({
   *   create: {
   *     // ... data to create a Favorite_champions
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the Favorite_champions we want to update
   *   }
   * })
   */
  upsert<T extends favorite_championsUpsertArgs>(args: Prisma.SelectSubset<T, favorite_championsUpsertArgs<ExtArgs>>): Prisma.Prisma__favorite_championsClient<runtime.Types.Result.GetResult<Prisma.$favorite_championsPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>


  /**
   * Count the number of Favorite_champions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {favorite_championsCountArgs} args - Arguments to filter Favorite_champions to count.
   * @example
   * // Count the number of Favorite_champions
   * const count = await prisma.favorite_champions.count({
   *   where: {
   *     // ... the filter for the Favorite_champions we want to count
   *   }
   * })
  **/
  count<T extends favorite_championsCountArgs>(
    args?: Prisma.Subset<T, favorite_championsCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], Favorite_championsCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a Favorite_champions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {Favorite_championsAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends Favorite_championsAggregateArgs>(args: Prisma.Subset<T, Favorite_championsAggregateArgs>): Prisma.PrismaPromise<GetFavorite_championsAggregateType<T>>

  /**
   * Group by Favorite_champions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {favorite_championsGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends favorite_championsGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: favorite_championsGroupByArgs['orderBy'] }
      : { orderBy?: favorite_championsGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, favorite_championsGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetFavorite_championsGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the favorite_champions model
 */
readonly fields: favorite_championsFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for favorite_champions.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__favorite_championsClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  champions<T extends Prisma.championsDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.championsDefaultArgs<ExtArgs>>): Prisma.Prisma__championsClient<runtime.Types.Result.GetResult<Prisma.$championsPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  players<T extends Prisma.playersDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.playersDefaultArgs<ExtArgs>>): Prisma.Prisma__playersClient<runtime.Types.Result.GetResult<Prisma.$playersPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the favorite_champions model
 */
export interface favorite_championsFieldRefs {
  readonly id: Prisma.FieldRef<"favorite_champions", 'Int'>
  readonly player_id: Prisma.FieldRef<"favorite_champions", 'Int'>
  readonly champion_id: Prisma.FieldRef<"favorite_champions", 'Int'>
  readonly added_at: Prisma.FieldRef<"favorite_champions", 'DateTime'>
}
    

// Custom InputTypes
/**
 * favorite_champions findUnique
 */
export type favorite_championsFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * Filter, which favorite_champions to fetch.
   */
  where: Prisma.favorite_championsWhereUniqueInput
}

/**
 * favorite_champions findUniqueOrThrow
 */
export type favorite_championsFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * Filter, which favorite_champions to fetch.
   */
  where: Prisma.favorite_championsWhereUniqueInput
}

/**
 * favorite_champions findFirst
 */
export type favorite_championsFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * Filter, which favorite_champions to fetch.
   */
  where?: Prisma.favorite_championsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of favorite_champions to fetch.
   */
  orderBy?: Prisma.favorite_championsOrderByWithRelationInput | Prisma.favorite_championsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for favorite_champions.
   */
  cursor?: Prisma.favorite_championsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` favorite_champions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` favorite_champions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of favorite_champions.
   */
  distinct?: Prisma.Favorite_championsScalarFieldEnum | Prisma.Favorite_championsScalarFieldEnum[]
}

/**
 * favorite_champions findFirstOrThrow
 */
export type favorite_championsFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * Filter, which favorite_champions to fetch.
   */
  where?: Prisma.favorite_championsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of favorite_champions to fetch.
   */
  orderBy?: Prisma.favorite_championsOrderByWithRelationInput | Prisma.favorite_championsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for favorite_champions.
   */
  cursor?: Prisma.favorite_championsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` favorite_champions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` favorite_champions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of favorite_champions.
   */
  distinct?: Prisma.Favorite_championsScalarFieldEnum | Prisma.Favorite_championsScalarFieldEnum[]
}

/**
 * favorite_champions findMany
 */
export type favorite_championsFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * Filter, which favorite_champions to fetch.
   */
  where?: Prisma.favorite_championsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of favorite_champions to fetch.
   */
  orderBy?: Prisma.favorite_championsOrderByWithRelationInput | Prisma.favorite_championsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing favorite_champions.
   */
  cursor?: Prisma.favorite_championsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` favorite_champions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` favorite_champions.
   */
  skip?: number
  distinct?: Prisma.Favorite_championsScalarFieldEnum | Prisma.Favorite_championsScalarFieldEnum[]
}

/**
 * favorite_champions create
 */
export type favorite_championsCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * The data needed to create a favorite_champions.
   */
  data: Prisma.XOR<Prisma.favorite_championsCreateInput, Prisma.favorite_championsUncheckedCreateInput>
}

/**
 * favorite_champions createMany
 */
export type favorite_championsCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many favorite_champions.
   */
  data: Prisma.favorite_championsCreateManyInput | Prisma.favorite_championsCreateManyInput[]
  skipDuplicates?: boolean
}

/**
 * favorite_champions createManyAndReturn
 */
export type favorite_championsCreateManyAndReturnArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelectCreateManyAndReturn<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * The data used to create many favorite_champions.
   */
  data: Prisma.favorite_championsCreateManyInput | Prisma.favorite_championsCreateManyInput[]
  skipDuplicates?: boolean
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsIncludeCreateManyAndReturn<ExtArgs> | null
}

/**
 * favorite_champions update
 */
export type favorite_championsUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * The data needed to update a favorite_champions.
   */
  data: Prisma.XOR<Prisma.favorite_championsUpdateInput, Prisma.favorite_championsUncheckedUpdateInput>
  /**
   * Choose, which favorite_champions to update.
   */
  where: Prisma.favorite_championsWhereUniqueInput
}

/**
 * favorite_champions updateMany
 */
export type favorite_championsUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update favorite_champions.
   */
  data: Prisma.XOR<Prisma.favorite_championsUpdateManyMutationInput, Prisma.favorite_championsUncheckedUpdateManyInput>
  /**
   * Filter which favorite_champions to update
   */
  where?: Prisma.favorite_championsWhereInput
  /**
   * Limit how many favorite_champions to update.
   */
  limit?: number
}

/**
 * favorite_champions updateManyAndReturn
 */
export type favorite_championsUpdateManyAndReturnArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelectUpdateManyAndReturn<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * The data used to update favorite_champions.
   */
  data: Prisma.XOR<Prisma.favorite_championsUpdateManyMutationInput, Prisma.favorite_championsUncheckedUpdateManyInput>
  /**
   * Filter which favorite_champions to update
   */
  where?: Prisma.favorite_championsWhereInput
  /**
   * Limit how many favorite_champions to update.
   */
  limit?: number
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsIncludeUpdateManyAndReturn<ExtArgs> | null
}

/**
 * favorite_champions upsert
 */
export type favorite_championsUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * The filter to search for the favorite_champions to update in case it exists.
   */
  where: Prisma.favorite_championsWhereUniqueInput
  /**
   * In case the favorite_champions found by the `where` argument doesn't exist, create a new favorite_champions with this data.
   */
  create: Prisma.XOR<Prisma.favorite_championsCreateInput, Prisma.favorite_championsUncheckedCreateInput>
  /**
   * In case the favorite_champions was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.favorite_championsUpdateInput, Prisma.favorite_championsUncheckedUpdateInput>
}

/**
 * favorite_champions delete
 */
export type favorite_championsDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
  /**
   * Filter which favorite_champions to delete.
   */
  where: Prisma.favorite_championsWhereUniqueInput
}

/**
 * favorite_champions deleteMany
 */
export type favorite_championsDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which favorite_champions to delete
   */
  where?: Prisma.favorite_championsWhereInput
  /**
   * Limit how many favorite_champions to delete.
   */
  limit?: number
}

/**
 * favorite_champions without action
 */
export type favorite_championsDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the favorite_champions
   */
  select?: Prisma.favorite_championsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the favorite_champions
   */
  omit?: Prisma.favorite_championsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.favorite_championsInclude<ExtArgs> | null
}
