// @prisma/client/models/items.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `items` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/library"
import type * as $Enums from "../enums.ts"
import type * as Prisma from "../internal/prismaNamespace.ts"

/**
 * Model items
 * 
 */
export type itemsModel = runtime.Types.Result.DefaultSelection<Prisma.$itemsPayload>

export type AggregateItems = {
  _count: ItemsCountAggregateOutputType | null
  _avg: ItemsAvgAggregateOutputType | null
  _sum: ItemsSumAggregateOutputType | null
  _min: ItemsMinAggregateOutputType | null
  _max: ItemsMaxAggregateOutputType | null
}

export type ItemsAvgAggregateOutputType = {
  id: number | null
  gold: number | null
  ap_bonus: number | null
  other_bonus: number | null
}

export type ItemsSumAggregateOutputType = {
  id: number | null
  gold: number | null
  ap_bonus: number | null
  other_bonus: number | null
}

export type ItemsMinAggregateOutputType = {
  id: number | null
  name: string | null
  description: string | null
  gold: number | null
  ap_bonus: number | null
  other_bonus: number | null
  image_url: string | null
  riot_id: string | null
  created_at: Date | null
}

export type ItemsMaxAggregateOutputType = {
  id: number | null
  name: string | null
  description: string | null
  gold: number | null
  ap_bonus: number | null
  other_bonus: number | null
  image_url: string | null
  riot_id: string | null
  created_at: Date | null
}

export type ItemsCountAggregateOutputType = {
  id: number
  name: number
  description: number
  gold: number
  ap_bonus: number
  other_bonus: number
  image_url: number
  riot_id: number
  created_at: number
  _all: number
}


export type ItemsAvgAggregateInputType = {
  id?: true
  gold?: true
  ap_bonus?: true
  other_bonus?: true
}

export type ItemsSumAggregateInputType = {
  id?: true
  gold?: true
  ap_bonus?: true
  other_bonus?: true
}

export type ItemsMinAggregateInputType = {
  id?: true
  name?: true
  description?: true
  gold?: true
  ap_bonus?: true
  other_bonus?: true
  image_url?: true
  riot_id?: true
  created_at?: true
}

export type ItemsMaxAggregateInputType = {
  id?: true
  name?: true
  description?: true
  gold?: true
  ap_bonus?: true
  other_bonus?: true
  image_url?: true
  riot_id?: true
  created_at?: true
}

export type ItemsCountAggregateInputType = {
  id?: true
  name?: true
  description?: true
  gold?: true
  ap_bonus?: true
  other_bonus?: true
  image_url?: true
  riot_id?: true
  created_at?: true
  _all?: true
}

export type ItemsAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which items to aggregate.
   */
  where?: Prisma.itemsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of items to fetch.
   */
  orderBy?: Prisma.itemsOrderByWithRelationInput | Prisma.itemsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.itemsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` items from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` items.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned items
  **/
  _count?: true | ItemsCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to average
  **/
  _avg?: ItemsAvgAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to sum
  **/
  _sum?: ItemsSumAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: ItemsMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: ItemsMaxAggregateInputType
}

export type GetItemsAggregateType<T extends ItemsAggregateArgs> = {
      [P in keyof T & keyof AggregateItems]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateItems[P]>
    : Prisma.GetScalarType<T[P], AggregateItems[P]>
}




export type itemsGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.itemsWhereInput
  orderBy?: Prisma.itemsOrderByWithAggregationInput | Prisma.itemsOrderByWithAggregationInput[]
  by: Prisma.ItemsScalarFieldEnum[] | Prisma.ItemsScalarFieldEnum
  having?: Prisma.itemsScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: ItemsCountAggregateInputType | true
  _avg?: ItemsAvgAggregateInputType
  _sum?: ItemsSumAggregateInputType
  _min?: ItemsMinAggregateInputType
  _max?: ItemsMaxAggregateInputType
}

export type ItemsGroupByOutputType = {
  id: number
  name: string
  description: string | null
  gold: number | null
  ap_bonus: number | null
  other_bonus: number | null
  image_url: string | null
  riot_id: string | null
  created_at: Date
  _count: ItemsCountAggregateOutputType | null
  _avg: ItemsAvgAggregateOutputType | null
  _sum: ItemsSumAggregateOutputType | null
  _min: ItemsMinAggregateOutputType | null
  _max: ItemsMaxAggregateOutputType | null
}

type GetItemsGroupByPayload<T extends itemsGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<ItemsGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof ItemsGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], ItemsGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], ItemsGroupByOutputType[P]>
      }
    >
  >



export type itemsWhereInput = {
  AND?: Prisma.itemsWhereInput | Prisma.itemsWhereInput[]
  OR?: Prisma.itemsWhereInput[]
  NOT?: Prisma.itemsWhereInput | Prisma.itemsWhereInput[]
  id?: Prisma.IntFilter<"items"> | number
  name?: Prisma.StringFilter<"items"> | string
  description?: Prisma.StringNullableFilter<"items"> | string | null
  gold?: Prisma.IntNullableFilter<"items"> | number | null
  ap_bonus?: Prisma.FloatNullableFilter<"items"> | number | null
  other_bonus?: Prisma.FloatNullableFilter<"items"> | number | null
  image_url?: Prisma.StringNullableFilter<"items"> | string | null
  riot_id?: Prisma.StringNullableFilter<"items"> | string | null
  created_at?: Prisma.DateTimeFilter<"items"> | Date | string
  ap_build_items?: Prisma.Ap_build_itemsListRelationFilter
}

export type itemsOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  description?: Prisma.SortOrderInput | Prisma.SortOrder
  gold?: Prisma.SortOrderInput | Prisma.SortOrder
  ap_bonus?: Prisma.SortOrderInput | Prisma.SortOrder
  other_bonus?: Prisma.SortOrderInput | Prisma.SortOrder
  image_url?: Prisma.SortOrderInput | Prisma.SortOrder
  riot_id?: Prisma.SortOrderInput | Prisma.SortOrder
  created_at?: Prisma.SortOrder
  ap_build_items?: Prisma.ap_build_itemsOrderByRelationAggregateInput
}

export type itemsWhereUniqueInput = Prisma.AtLeast<{
  id?: number
  name?: string
  AND?: Prisma.itemsWhereInput | Prisma.itemsWhereInput[]
  OR?: Prisma.itemsWhereInput[]
  NOT?: Prisma.itemsWhereInput | Prisma.itemsWhereInput[]
  description?: Prisma.StringNullableFilter<"items"> | string | null
  gold?: Prisma.IntNullableFilter<"items"> | number | null
  ap_bonus?: Prisma.FloatNullableFilter<"items"> | number | null
  other_bonus?: Prisma.FloatNullableFilter<"items"> | number | null
  image_url?: Prisma.StringNullableFilter<"items"> | string | null
  riot_id?: Prisma.StringNullableFilter<"items"> | string | null
  created_at?: Prisma.DateTimeFilter<"items"> | Date | string
  ap_build_items?: Prisma.Ap_build_itemsListRelationFilter
}, "id" | "name">

export type itemsOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  description?: Prisma.SortOrderInput | Prisma.SortOrder
  gold?: Prisma.SortOrderInput | Prisma.SortOrder
  ap_bonus?: Prisma.SortOrderInput | Prisma.SortOrder
  other_bonus?: Prisma.SortOrderInput | Prisma.SortOrder
  image_url?: Prisma.SortOrderInput | Prisma.SortOrder
  riot_id?: Prisma.SortOrderInput | Prisma.SortOrder
  created_at?: Prisma.SortOrder
  _count?: Prisma.itemsCountOrderByAggregateInput
  _avg?: Prisma.itemsAvgOrderByAggregateInput
  _max?: Prisma.itemsMaxOrderByAggregateInput
  _min?: Prisma.itemsMinOrderByAggregateInput
  _sum?: Prisma.itemsSumOrderByAggregateInput
}

export type itemsScalarWhereWithAggregatesInput = {
  AND?: Prisma.itemsScalarWhereWithAggregatesInput | Prisma.itemsScalarWhereWithAggregatesInput[]
  OR?: Prisma.itemsScalarWhereWithAggregatesInput[]
  NOT?: Prisma.itemsScalarWhereWithAggregatesInput | Prisma.itemsScalarWhereWithAggregatesInput[]
  id?: Prisma.IntWithAggregatesFilter<"items"> | number
  name?: Prisma.StringWithAggregatesFilter<"items"> | string
  description?: Prisma.StringNullableWithAggregatesFilter<"items"> | string | null
  gold?: Prisma.IntNullableWithAggregatesFilter<"items"> | number | null
  ap_bonus?: Prisma.FloatNullableWithAggregatesFilter<"items"> | number | null
  other_bonus?: Prisma.FloatNullableWithAggregatesFilter<"items"> | number | null
  image_url?: Prisma.StringNullableWithAggregatesFilter<"items"> | string | null
  riot_id?: Prisma.StringNullableWithAggregatesFilter<"items"> | string | null
  created_at?: Prisma.DateTimeWithAggregatesFilter<"items"> | Date | string
}

export type itemsCreateInput = {
  name: string
  description?: string | null
  gold?: number | null
  ap_bonus?: number | null
  other_bonus?: number | null
  image_url?: string | null
  riot_id?: string | null
  created_at: Date | string
  ap_build_items?: Prisma.ap_build_itemsCreateNestedManyWithoutItemsInput
}

export type itemsUncheckedCreateInput = {
  id?: number
  name: string
  description?: string | null
  gold?: number | null
  ap_bonus?: number | null
  other_bonus?: number | null
  image_url?: string | null
  riot_id?: string | null
  created_at: Date | string
  ap_build_items?: Prisma.ap_build_itemsUncheckedCreateNestedManyWithoutItemsInput
}

export type itemsUpdateInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  gold?: Prisma.NullableIntFieldUpdateOperationsInput | number | null
  ap_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  other_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  ap_build_items?: Prisma.ap_build_itemsUpdateManyWithoutItemsNestedInput
}

export type itemsUncheckedUpdateInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  name?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  gold?: Prisma.NullableIntFieldUpdateOperationsInput | number | null
  ap_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  other_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  ap_build_items?: Prisma.ap_build_itemsUncheckedUpdateManyWithoutItemsNestedInput
}

export type itemsCreateManyInput = {
  id?: number
  name: string
  description?: string | null
  gold?: number | null
  ap_bonus?: number | null
  other_bonus?: number | null
  image_url?: string | null
  riot_id?: string | null
  created_at: Date | string
}

export type itemsUpdateManyMutationInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  gold?: Prisma.NullableIntFieldUpdateOperationsInput | number | null
  ap_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  other_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type itemsUncheckedUpdateManyInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  name?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  gold?: Prisma.NullableIntFieldUpdateOperationsInput | number | null
  ap_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  other_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ItemsScalarRelationFilter = {
  is?: Prisma.itemsWhereInput
  isNot?: Prisma.itemsWhereInput
}

export type itemsCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  description?: Prisma.SortOrder
  gold?: Prisma.SortOrder
  ap_bonus?: Prisma.SortOrder
  other_bonus?: Prisma.SortOrder
  image_url?: Prisma.SortOrder
  riot_id?: Prisma.SortOrder
  created_at?: Prisma.SortOrder
}

export type itemsAvgOrderByAggregateInput = {
  id?: Prisma.SortOrder
  gold?: Prisma.SortOrder
  ap_bonus?: Prisma.SortOrder
  other_bonus?: Prisma.SortOrder
}

export type itemsMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  description?: Prisma.SortOrder
  gold?: Prisma.SortOrder
  ap_bonus?: Prisma.SortOrder
  other_bonus?: Prisma.SortOrder
  image_url?: Prisma.SortOrder
  riot_id?: Prisma.SortOrder
  created_at?: Prisma.SortOrder
}

export type itemsMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  description?: Prisma.SortOrder
  gold?: Prisma.SortOrder
  ap_bonus?: Prisma.SortOrder
  other_bonus?: Prisma.SortOrder
  image_url?: Prisma.SortOrder
  riot_id?: Prisma.SortOrder
  created_at?: Prisma.SortOrder
}

export type itemsSumOrderByAggregateInput = {
  id?: Prisma.SortOrder
  gold?: Prisma.SortOrder
  ap_bonus?: Prisma.SortOrder
  other_bonus?: Prisma.SortOrder
}

export type itemsCreateNestedOneWithoutAp_build_itemsInput = {
  create?: Prisma.XOR<Prisma.itemsCreateWithoutAp_build_itemsInput, Prisma.itemsUncheckedCreateWithoutAp_build_itemsInput>
  connectOrCreate?: Prisma.itemsCreateOrConnectWithoutAp_build_itemsInput
  connect?: Prisma.itemsWhereUniqueInput
}

export type itemsUpdateOneRequiredWithoutAp_build_itemsNestedInput = {
  create?: Prisma.XOR<Prisma.itemsCreateWithoutAp_build_itemsInput, Prisma.itemsUncheckedCreateWithoutAp_build_itemsInput>
  connectOrCreate?: Prisma.itemsCreateOrConnectWithoutAp_build_itemsInput
  upsert?: Prisma.itemsUpsertWithoutAp_build_itemsInput
  connect?: Prisma.itemsWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.itemsUpdateToOneWithWhereWithoutAp_build_itemsInput, Prisma.itemsUpdateWithoutAp_build_itemsInput>, Prisma.itemsUncheckedUpdateWithoutAp_build_itemsInput>
}

export type itemsCreateWithoutAp_build_itemsInput = {
  name: string
  description?: string | null
  gold?: number | null
  ap_bonus?: number | null
  other_bonus?: number | null
  image_url?: string | null
  riot_id?: string | null
  created_at: Date | string
}

export type itemsUncheckedCreateWithoutAp_build_itemsInput = {
  id?: number
  name: string
  description?: string | null
  gold?: number | null
  ap_bonus?: number | null
  other_bonus?: number | null
  image_url?: string | null
  riot_id?: string | null
  created_at: Date | string
}

export type itemsCreateOrConnectWithoutAp_build_itemsInput = {
  where: Prisma.itemsWhereUniqueInput
  create: Prisma.XOR<Prisma.itemsCreateWithoutAp_build_itemsInput, Prisma.itemsUncheckedCreateWithoutAp_build_itemsInput>
}

export type itemsUpsertWithoutAp_build_itemsInput = {
  update: Prisma.XOR<Prisma.itemsUpdateWithoutAp_build_itemsInput, Prisma.itemsUncheckedUpdateWithoutAp_build_itemsInput>
  create: Prisma.XOR<Prisma.itemsCreateWithoutAp_build_itemsInput, Prisma.itemsUncheckedCreateWithoutAp_build_itemsInput>
  where?: Prisma.itemsWhereInput
}

export type itemsUpdateToOneWithWhereWithoutAp_build_itemsInput = {
  where?: Prisma.itemsWhereInput
  data: Prisma.XOR<Prisma.itemsUpdateWithoutAp_build_itemsInput, Prisma.itemsUncheckedUpdateWithoutAp_build_itemsInput>
}

export type itemsUpdateWithoutAp_build_itemsInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  gold?: Prisma.NullableIntFieldUpdateOperationsInput | number | null
  ap_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  other_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type itemsUncheckedUpdateWithoutAp_build_itemsInput = {
  id?: Prisma.IntFieldUpdateOperationsInput | number
  name?: Prisma.StringFieldUpdateOperationsInput | string
  description?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  gold?: Prisma.NullableIntFieldUpdateOperationsInput | number | null
  ap_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  other_bonus?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  image_url?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  riot_id?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  created_at?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}


/**
 * Count Type ItemsCountOutputType
 */

export type ItemsCountOutputType = {
  ap_build_items: number
}

export type ItemsCountOutputTypeSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  ap_build_items?: boolean | ItemsCountOutputTypeCountAp_build_itemsArgs
}

/**
 * ItemsCountOutputType without action
 */
export type ItemsCountOutputTypeDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ItemsCountOutputType
   */
  select?: Prisma.ItemsCountOutputTypeSelect<ExtArgs> | null
}

/**
 * ItemsCountOutputType without action
 */
export type ItemsCountOutputTypeCountAp_build_itemsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.ap_build_itemsWhereInput
}


export type itemsSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  name?: boolean
  description?: boolean
  gold?: boolean
  ap_bonus?: boolean
  other_bonus?: boolean
  image_url?: boolean
  riot_id?: boolean
  created_at?: boolean
  ap_build_items?: boolean | Prisma.items$ap_build_itemsArgs<ExtArgs>
  _count?: boolean | Prisma.ItemsCountOutputTypeDefaultArgs<ExtArgs>
}, ExtArgs["result"]["items"]>

export type itemsSelectCreateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  name?: boolean
  description?: boolean
  gold?: boolean
  ap_bonus?: boolean
  other_bonus?: boolean
  image_url?: boolean
  riot_id?: boolean
  created_at?: boolean
}, ExtArgs["result"]["items"]>

export type itemsSelectUpdateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  name?: boolean
  description?: boolean
  gold?: boolean
  ap_bonus?: boolean
  other_bonus?: boolean
  image_url?: boolean
  riot_id?: boolean
  created_at?: boolean
}, ExtArgs["result"]["items"]>

export type itemsSelectScalar = {
  id?: boolean
  name?: boolean
  description?: boolean
  gold?: boolean
  ap_bonus?: boolean
  other_bonus?: boolean
  image_url?: boolean
  riot_id?: boolean
  created_at?: boolean
}

export type itemsOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "name" | "description" | "gold" | "ap_bonus" | "other_bonus" | "image_url" | "riot_id" | "created_at", ExtArgs["result"]["items"]>
export type itemsInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  ap_build_items?: boolean | Prisma.items$ap_build_itemsArgs<ExtArgs>
  _count?: boolean | Prisma.ItemsCountOutputTypeDefaultArgs<ExtArgs>
}
export type itemsIncludeCreateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {}
export type itemsIncludeUpdateManyAndReturn<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {}

export type $itemsPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "items"
  objects: {
    ap_build_items: Prisma.$ap_build_itemsPayload<ExtArgs>[]
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: number
    name: string
    description: string | null
    gold: number | null
    ap_bonus: number | null
    other_bonus: number | null
    image_url: string | null
    riot_id: string | null
    created_at: Date
  }, ExtArgs["result"]["items"]>
  composites: {}
}

export type itemsGetPayload<S extends boolean | null | undefined | itemsDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$itemsPayload, S>

export type itemsCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<itemsFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: ItemsCountAggregateInputType | true
  }

export interface itemsDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['items'], meta: { name: 'items' } }
  /**
   * Find zero or one Items that matches the filter.
   * @param {itemsFindUniqueArgs} args - Arguments to find a Items
   * @example
   * // Get one Items
   * const items = await prisma.items.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends itemsFindUniqueArgs>(args: Prisma.SelectSubset<T, itemsFindUniqueArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one Items that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {itemsFindUniqueOrThrowArgs} args - Arguments to find a Items
   * @example
   * // Get one Items
   * const items = await prisma.items.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends itemsFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, itemsFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Items that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {itemsFindFirstArgs} args - Arguments to find a Items
   * @example
   * // Get one Items
   * const items = await prisma.items.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends itemsFindFirstArgs>(args?: Prisma.SelectSubset<T, itemsFindFirstArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Items that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {itemsFindFirstOrThrowArgs} args - Arguments to find a Items
   * @example
   * // Get one Items
   * const items = await prisma.items.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends itemsFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, itemsFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Items that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {itemsFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all Items
   * const items = await prisma.items.findMany()
   * 
   * // Get first 10 Items
   * const items = await prisma.items.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const itemsWithIdOnly = await prisma.items.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends itemsFindManyArgs>(args?: Prisma.SelectSubset<T, itemsFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a Items.
   * @param {itemsCreateArgs} args - Arguments to create a Items.
   * @example
   * // Create one Items
   * const Items = await prisma.items.create({
   *   data: {
   *     // ... data to create a Items
   *   }
   * })
   * 
   */
  create<T extends itemsCreateArgs>(args: Prisma.SelectSubset<T, itemsCreateArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many Items.
   * @param {itemsCreateManyArgs} args - Arguments to create many Items.
   * @example
   * // Create many Items
   * const items = await prisma.items.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends itemsCreateManyArgs>(args?: Prisma.SelectSubset<T, itemsCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create many Items and returns the data saved in the database.
   * @param {itemsCreateManyAndReturnArgs} args - Arguments to create many Items.
   * @example
   * // Create many Items
   * const items = await prisma.items.createManyAndReturn({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * 
   * // Create many Items and only return the `id`
   * const itemsWithIdOnly = await prisma.items.createManyAndReturn({
   *   select: { id: true },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * 
   */
  createManyAndReturn<T extends itemsCreateManyAndReturnArgs>(args?: Prisma.SelectSubset<T, itemsCreateManyAndReturnArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "createManyAndReturn", GlobalOmitOptions>>

  /**
   * Delete a Items.
   * @param {itemsDeleteArgs} args - Arguments to delete one Items.
   * @example
   * // Delete one Items
   * const Items = await prisma.items.delete({
   *   where: {
   *     // ... filter to delete one Items
   *   }
   * })
   * 
   */
  delete<T extends itemsDeleteArgs>(args: Prisma.SelectSubset<T, itemsDeleteArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one Items.
   * @param {itemsUpdateArgs} args - Arguments to update one Items.
   * @example
   * // Update one Items
   * const items = await prisma.items.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends itemsUpdateArgs>(args: Prisma.SelectSubset<T, itemsUpdateArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more Items.
   * @param {itemsDeleteManyArgs} args - Arguments to filter Items to delete.
   * @example
   * // Delete a few Items
   * const { count } = await prisma.items.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends itemsDeleteManyArgs>(args?: Prisma.SelectSubset<T, itemsDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Items.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {itemsUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many Items
   * const items = await prisma.items.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends itemsUpdateManyArgs>(args: Prisma.SelectSubset<T, itemsUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Items and returns the data updated in the database.
   * @param {itemsUpdateManyAndReturnArgs} args - Arguments to update many Items.
   * @example
   * // Update many Items
   * const items = await prisma.items.updateManyAndReturn({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * 
   * // Update zero or more Items and only return the `id`
   * const itemsWithIdOnly = await prisma.items.updateManyAndReturn({
   *   select: { id: true },
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * 
   */
  updateManyAndReturn<T extends itemsUpdateManyAndReturnArgs>(args: Prisma.SelectSubset<T, itemsUpdateManyAndReturnArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "updateManyAndReturn", GlobalOmitOptions>>

  /**
   * Create or update one Items.
   * @param {itemsUpsertArgs} args - Arguments to update or create a Items.
   * @example
   * // Update or create a Items
   * const items = await prisma.items.upsert({
   *   create: {
   *     // ... data to create a Items
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the Items we want to update
   *   }
   * })
   */
  upsert<T extends itemsUpsertArgs>(args: Prisma.SelectSubset<T, itemsUpsertArgs<ExtArgs>>): Prisma.Prisma__itemsClient<runtime.Types.Result.GetResult<Prisma.$itemsPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>


  /**
   * Count the number of Items.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {itemsCountArgs} args - Arguments to filter Items to count.
   * @example
   * // Count the number of Items
   * const count = await prisma.items.count({
   *   where: {
   *     // ... the filter for the Items we want to count
   *   }
   * })
  **/
  count<T extends itemsCountArgs>(
    args?: Prisma.Subset<T, itemsCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], ItemsCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a Items.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ItemsAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends ItemsAggregateArgs>(args: Prisma.Subset<T, ItemsAggregateArgs>): Prisma.PrismaPromise<GetItemsAggregateType<T>>

  /**
   * Group by Items.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {itemsGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends itemsGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: itemsGroupByArgs['orderBy'] }
      : { orderBy?: itemsGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, itemsGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetItemsGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the items model
 */
readonly fields: itemsFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for items.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__itemsClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  ap_build_items<T extends Prisma.items$ap_build_itemsArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.items$ap_build_itemsArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ap_build_itemsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the items model
 */
export interface itemsFieldRefs {
  readonly id: Prisma.FieldRef<"items", 'Int'>
  readonly name: Prisma.FieldRef<"items", 'String'>
  readonly description: Prisma.FieldRef<"items", 'String'>
  readonly gold: Prisma.FieldRef<"items", 'Int'>
  readonly ap_bonus: Prisma.FieldRef<"items", 'Float'>
  readonly other_bonus: Prisma.FieldRef<"items", 'Float'>
  readonly image_url: Prisma.FieldRef<"items", 'String'>
  readonly riot_id: Prisma.FieldRef<"items", 'String'>
  readonly created_at: Prisma.FieldRef<"items", 'DateTime'>
}
    

// Custom InputTypes
/**
 * items findUnique
 */
export type itemsFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * Filter, which items to fetch.
   */
  where: Prisma.itemsWhereUniqueInput
}

/**
 * items findUniqueOrThrow
 */
export type itemsFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * Filter, which items to fetch.
   */
  where: Prisma.itemsWhereUniqueInput
}

/**
 * items findFirst
 */
export type itemsFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * Filter, which items to fetch.
   */
  where?: Prisma.itemsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of items to fetch.
   */
  orderBy?: Prisma.itemsOrderByWithRelationInput | Prisma.itemsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for items.
   */
  cursor?: Prisma.itemsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` items from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` items.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of items.
   */
  distinct?: Prisma.ItemsScalarFieldEnum | Prisma.ItemsScalarFieldEnum[]
}

/**
 * items findFirstOrThrow
 */
export type itemsFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * Filter, which items to fetch.
   */
  where?: Prisma.itemsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of items to fetch.
   */
  orderBy?: Prisma.itemsOrderByWithRelationInput | Prisma.itemsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for items.
   */
  cursor?: Prisma.itemsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` items from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` items.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of items.
   */
  distinct?: Prisma.ItemsScalarFieldEnum | Prisma.ItemsScalarFieldEnum[]
}

/**
 * items findMany
 */
export type itemsFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * Filter, which items to fetch.
   */
  where?: Prisma.itemsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of items to fetch.
   */
  orderBy?: Prisma.itemsOrderByWithRelationInput | Prisma.itemsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing items.
   */
  cursor?: Prisma.itemsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` items from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` items.
   */
  skip?: number
  distinct?: Prisma.ItemsScalarFieldEnum | Prisma.ItemsScalarFieldEnum[]
}

/**
 * items create
 */
export type itemsCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * The data needed to create a items.
   */
  data: Prisma.XOR<Prisma.itemsCreateInput, Prisma.itemsUncheckedCreateInput>
}

/**
 * items createMany
 */
export type itemsCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many items.
   */
  data: Prisma.itemsCreateManyInput | Prisma.itemsCreateManyInput[]
  skipDuplicates?: boolean
}

/**
 * items createManyAndReturn
 */
export type itemsCreateManyAndReturnArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelectCreateManyAndReturn<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * The data used to create many items.
   */
  data: Prisma.itemsCreateManyInput | Prisma.itemsCreateManyInput[]
  skipDuplicates?: boolean
}

/**
 * items update
 */
export type itemsUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * The data needed to update a items.
   */
  data: Prisma.XOR<Prisma.itemsUpdateInput, Prisma.itemsUncheckedUpdateInput>
  /**
   * Choose, which items to update.
   */
  where: Prisma.itemsWhereUniqueInput
}

/**
 * items updateMany
 */
export type itemsUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update items.
   */
  data: Prisma.XOR<Prisma.itemsUpdateManyMutationInput, Prisma.itemsUncheckedUpdateManyInput>
  /**
   * Filter which items to update
   */
  where?: Prisma.itemsWhereInput
  /**
   * Limit how many items to update.
   */
  limit?: number
}

/**
 * items updateManyAndReturn
 */
export type itemsUpdateManyAndReturnArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelectUpdateManyAndReturn<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * The data used to update items.
   */
  data: Prisma.XOR<Prisma.itemsUpdateManyMutationInput, Prisma.itemsUncheckedUpdateManyInput>
  /**
   * Filter which items to update
   */
  where?: Prisma.itemsWhereInput
  /**
   * Limit how many items to update.
   */
  limit?: number
}

/**
 * items upsert
 */
export type itemsUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * The filter to search for the items to update in case it exists.
   */
  where: Prisma.itemsWhereUniqueInput
  /**
   * In case the items found by the `where` argument doesn't exist, create a new items with this data.
   */
  create: Prisma.XOR<Prisma.itemsCreateInput, Prisma.itemsUncheckedCreateInput>
  /**
   * In case the items was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.itemsUpdateInput, Prisma.itemsUncheckedUpdateInput>
}

/**
 * items delete
 */
export type itemsDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
  /**
   * Filter which items to delete.
   */
  where: Prisma.itemsWhereUniqueInput
}

/**
 * items deleteMany
 */
export type itemsDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which items to delete
   */
  where?: Prisma.itemsWhereInput
  /**
   * Limit how many items to delete.
   */
  limit?: number
}

/**
 * items.ap_build_items
 */
export type items$ap_build_itemsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ap_build_items
   */
  select?: Prisma.ap_build_itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ap_build_items
   */
  omit?: Prisma.ap_build_itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ap_build_itemsInclude<ExtArgs> | null
  where?: Prisma.ap_build_itemsWhereInput
  orderBy?: Prisma.ap_build_itemsOrderByWithRelationInput | Prisma.ap_build_itemsOrderByWithRelationInput[]
  cursor?: Prisma.ap_build_itemsWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.Ap_build_itemsScalarFieldEnum | Prisma.Ap_build_itemsScalarFieldEnum[]
}

/**
 * items without action
 */
export type itemsDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the items
   */
  select?: Prisma.itemsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the items
   */
  omit?: Prisma.itemsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.itemsInclude<ExtArgs> | null
}
